// gh-series/tests/util.js
/**
 * Copyright (c) 2015 "Fronteer LTD"
 * Grasshopper Event Engine
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

var _ = require('lodash');
var assert = require('assert');
var moment = require('moment');

var Counter = require('gh-tests/lib/counter');
var EventsTestsUtil = require('gh-events/tests/util');
var UsersTestsUtil = require('gh-users/tests/util');
var TestsUtil = require('gh-tests');

var Listener = require('gh-series/lib/internal/patterns/cambridge/listener');

// Keep track of the asynchronous operations when generating
// the metadata if events are added/removed to a serie
var serieEventsCounter = new Counter();

Listener.on('preGeneration', function() { serieEventsCounter.incr(); });
Listener.on('postGeneration', function() { serieEventsCounter.decr(); });

/**
 * Assert that a serie has all expected properties
 *
 * @param  {Serie}              serie               The serie to assert the properties for
 * @param  {Serie}              expectedSerie       The serie to which the provided serie should be compared
 * @throws {AssertionError}                         Error thrown when an assertion failed
 */
var assertSerie = module.exports.assertSerie = function(serie, expectedSerie) {
    assert.ok(serie);
    assert.ok(serie.id);
    assert.ok(serie.AppId);
    assert.ok(serie.displayName);
    assert.ok(serie.createdAt);
    assert.ok(serie.updatedAt);
    assert.strictEqual(serie.id, expectedSerie.id);
    assert.strictEqual(serie.AppId, expectedSerie.AppId);
    assert.strictEqual(serie.displayName, expectedSerie.displayName);
    assert.strictEqual(serie.description, expectedSerie.description);
    assert.strictEqual(serie.createdAt, expectedSerie.createdAt);
    assert.strictEqual(serie.GroupId, expectedSerie.GroupId);
};

/**
 * Assert that a new serie can be created
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {String}             displayName                     The name of the serie
 * @param  {Object}             [opts]                          A set of optional parameters
 * @param  {Number}             [opts.app]                      The id of the application to which this serie will belong
 * @param  {String}             [opts.description]              The description of the serie
 * @param  {Number}             [opts.group]                    The id of the group that can manage the serie
 * @param  {Function}           callback                        Standard callback function
 * @param  {Serie}              callback.serie                  The created serie
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertCreateSerie = module.exports.assertCreateSerie = function(client, displayName, opts, callback) {
    opts = opts || {};

    client.serie.createSerie(displayName, opts, function(err, createdSerie, response) {
        assert.ok(!err);
        assert.ok(createdSerie);
        assert.ok(createdSerie.id);
        assert.ok(createdSerie.createdAt);
        assert.ok(createdSerie.updatedAt);
        assert.strictEqual(createdSerie.displayName, displayName);
        if (opts.app) {
            assert.strictEqual(createdSerie.AppId, opts.app);
        }
        if (opts.description) {
            assert.strictEqual(createdSerie.description, opts.description);
        }
        if (opts.group) {
            assert.strictEqual(createdSerie.GroupId, opts.group);
        }
        return callback(createdSerie);
    });
};

/**
 * Assert that a serie can not be created
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {String}             displayName                     The name of the serie
 * @param  {Object}             [opts]                          A set of optional parameters
 * @param  {Number}             [opts.app]                      The id of the application to which this serie will belong
 * @param  {String}             [opts.description]              The description of the serie
 * @param  {Number}             [opts.group]                    The id of the group that can manage the serie
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertCreateSerieFails = module.exports.assertCreateSerieFails = function(client, displayName, opts, code, callback) {
    client.serie.createSerie(displayName, opts, function(err, serie, response) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        assert.ok(!serie);

        return callback();
    });
};

/**
 * Assert that a series can be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the series to retrieve
 * @param  {Boolean}            [includeOrgUnits]               Whether the organisational units that this series is a part of should be included in the response. Defaults to `false`
 * @param  {Boolean}            [expectedCanManage]             The expected value for the `canManage` flag
 * @param  {Serie}              [expectedSeries]                The expected series to be retrieved
 * @param  {Function}           callback                        Standard callback function
 * @param  {Serie}              callback.series                 The retrieved series
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetSeries = module.exports.assertGetSeries = function(client, id, includeOrgUnits, expectedCanManage, expectedSeries, callback) {
    // Wait until all series and/or event updates have been dealt with
    serieEventsCounter.whenZero(function() {

        // Get the series
        client.serie.getSeries(id, includeOrgUnits, function(err, series) {
            assert.ok(!err);
            assert.ok(series);
            assert.strictEqual(series.id, id);

            // The canManage flag should always be returned
            assert.ok(_.isBoolean(series.canManage));
            assert.strictEqual(series.canManage, expectedCanManage);

            if (includeOrgUnits) {
                assert.ok(_.isArray(series.OrgUnits));
            }
            if (expectedSeries) {
                assertSerie(series, expectedSeries);
            }
            return callback(series);
        });
    });
};

/**
 * Assert that a series can not be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the series to retrieve
 * @param  {Boolean}            includeOrgUnits                 Whether the organisational units that this series is a part of should be included in the response
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetSeriesFails = module.exports.assertGetSeriesFails = function(client, id, includeOrgUnits, code, callback) {
    client.serie.getSeries(id, includeOrgUnits, function(err, series) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        assert.ok(!series);
        return callback();
    });
};

/**
 * Assert that a serie can be updated
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie to update
 * @param  {Object}             update                          The updates to persist
 * @param  {Function}           callback                        Standard callback function
 * @param  {Serie}              callback.serie                  The updated serie
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertUpdateSerie = module.exports.assertUpdateSerie = function(client, id, update, callback) {
    client.serie.updateSerie(id, update, function(err, serie) {
        assert.ok(!err);
        assert.ok(serie);
        assert.strictEqual(serie.id, id);
        if (update.displayName) {
            assert.strictEqual(serie.displayName, update.displayName);
        }
        if (update.description) {
            assert.strictEqual(serie.description, update.description);
        }
        if (update.group) {
            assert.strictEqual(serie.GroupId, update.group);
        }
        return callback(serie);
    });
};

/**
 * Assert that an serie can not be updated
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie to update
 * @param  {Object}             update                          The updates to persist
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertUpdateSerieFails = module.exports.assertUpdateSerieFails = function(client, id, update, code, callback) {
    client.serie.updateSerie(id, update, function(err, serie) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        assert.ok(!serie);
        return callback();
    });
};

/**
 * Assert that a serie can be deleted
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie to delete
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertDeleteSerie = module.exports.assertDeleteSerie = function(client, id, callback) {
    client.serie.deleteSerie(id, function(err) {
        assert.ok(!err);
        return callback();
    });
};

/**
 * Assert that a serie can not be deleted
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie to delete
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertDeleteSerieFails = module.exports.assertDeleteSerieFails = function(client, id, code, callback) {
    client.serie.deleteSerie(id, function(err) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        return callback();
    });
};

/**
 * Assert that events can be added to a serie
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie to add events to
 * @param  {Number[]}           events                          The id(s) of the event(s) to add to a serie
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertAddSeriesEvents = module.exports.assertAddSeriesEvents = function(client, id, events, callback) {
    client.serie.addSeriesEvents(id, events, function(err) {
        assert.ok(!err);
        serieEventsCounter.whenZero(callback);
    });
};

/**
 * Assert that events can not be added to a serie
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie to add events to
 * @param  {Number[]}           events                          The id(s) of the event(s) to add to a serie
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertAddSeriesEventsFails = module.exports.assertAddSeriesEventsFails = function(client, id, events, code, callback) {
    client.serie.addSeriesEvents(id, events, function(err) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        return callback();
    });
};

/**
 * Assert that events can be removed from a serie
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie to remove events from
 * @param  {Number[]}           events                          The id(s) of the event(s) to remove from a serie
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertDeleteSeriesEvents = module.exports.assertDeleteSeriesEvents = function(client, id, events, callback) {
    client.serie.deleteSeriesEvents(id, events, function(err) {
        assert.ok(!err);
        serieEventsCounter.whenZero(callback);
    });
};

/**
 * Assert that events can not be removed from a serie
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie to remove events from
 * @param  {Number[]}           events                          The id(s) of the event(s) to remove from a serie
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertDeleteSeriesEventsFails = module.exports.assertDeleteSeriesEventsFails = function(client, id, events, code, callback) {
    client.serie.deleteSeriesEvents(id, events, function(err) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        return callback();
    });
};

/**
 * Assert that the events for a serie can be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie for wich to get the events
 * @param  {Number}             [limit]                         The number of events that should be retrieved, defaults to 10
 * @param  {Number}             [offset]                        The number to start paging from, defaults to 0
 * @param  {Boolean}            [upcoming]                      Whether to only include upcoming events. Defaults to `true`
 * @param  {Number[]}           [expectedEvents]                The id(s) of the event(s) that should be in this calendar
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetSeriesEvents = module.exports.assertGetSeriesEvents = function(client, id, limit, offset, upcoming, expectedEvents, callback) {
    client.serie.getSeriesEvents(id, limit, offset, upcoming, function(err, events) {
        assert.ok(!err);
        assert.ok(events);
        assert.ok(events.results);

        // Assert that the events are sorted based on their start times
        for (var i = 1; i < events.results.length; i++) {
            assert.ok(events.results[i].start >= events.results[i - 1].start);
        }

        // Assert that the returned events are the ones we expected
        if (expectedEvents) {
            assert.strictEqual(events.results.length, expectedEvents.length);
            _.each(events.results, function(event, i) {
                assert.strictEqual(event.id, expectedEvents[i].id);

                // Assert each event includes the organisers
                assert.ok(event.organisers);
            });
        }

        return callback(events);
    });
};

/**
 * Assert that the events for a serie can not be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie for wich to get the events
 * @param  {Number}             [limit]                         The number of events that should be retrieved, defaults to 10
 * @param  {Number}             [offset]                        The number to start paging from, defaults to 0
 * @param  {Boolean}            [upcoming]                      Whether to only include upcoming events. Defaults to `true`
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetSeriesEventsFails = module.exports.assertGetSeriesEventsFails = function(client, id, limit, offset, upcoming, code, callback) {
    client.serie.getSeriesEvents(id, limit, offset, upcoming, function(err, events) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        return callback();
    });
};

/**
 * Assert that the JSON event calendar for a serie can be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie for wich to get the event calendar
 * @param  {String}             start                           The start date for the calendar
 * @param  {String}             end                             The end date for the calendar
 * @param  {Event[]}            [expectedEvents]                The event(s) that should be in this calendar
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetSeriesCalendar = module.exports.assertGetSeriesCalendar = function(client, id, start, end, expectedEvents, callback) {
    client.serie.getSeriesCalendar(id, start, end, function(err, calendar, response) {
        assert.ok(!err);

        // Assert the response is JSON
        assert.strictEqual(response.headers['content-type'], 'application/json; charset=utf-8');

        // Assert the returned results
        assert.ok(calendar);
        assert.ok(calendar.results);

        // Assert that the events are sorted based on their start times
        for (var i = 1; i < calendar.results.length; i++) {
            assert.ok(calendar.results[i].start > calendar.results[i - 1].start);
        }

        // Assert that the returned events are the ones we expected
        if (expectedEvents) {
            assert.strictEqual(calendar.results.length, expectedEvents.length);
            _.each(calendar.results, function(event, i) {
                assert.strictEqual(event.id, expectedEvents[i].id);
            });
        }

        return callback(calendar);
    });
};

/**
 * Assert that the JSON event calendar for a serie can not be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie for wich to get the event calendar
 * @param  {String}             start                           The start date for the calendar
 * @param  {String}             end                             The end date for the calendar
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetSeriesCalendarFails = module.exports.assertGetSeriesCalendarFails = function(client, id, start, end, code, callback) {
    client.serie.getSeriesCalendar(id, start, end, function(err, calendar) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        assert.ok(!calendar);
        return callback();
    });
};

/**
 * Assert that the RSS event calendar for a serie can be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie for wich to get the event calendar
 * @param  {String}             start                           The start date for the calendar
 * @param  {String}             end                             The end date for the calendar
 * @param  {Event[]}            [expectedEvents]                The event(s) that should be in this calendar
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetSeriesCalendarRss = module.exports.assertGetSeriesCalendarRss = function(client, id, start, end, expectedEvents, callback) {
    client.serie.getSeriesCalendarRss(id, start, end, function(err, calendar, response) {
        assert.ok(!err);

        // Assert the response is RSS
        assert.strictEqual(response.headers['content-type'], 'application/rss+xml; charset=utf-8');

        // Parse the RSS calendar
        TestsUtil.parseRssCalendar(calendar, function(calendar) {

            // Check if the returned events are the ones we expected
            if (expectedEvents) {
                if (_.isEmpty(expectedEvents)) {
                    // If no items should be returned, the RSS feed will not contain an `<item>` tag
                    assert.ok(!calendar.item);
                } else {
                    assert.strictEqual(calendar.item.length, expectedEvents.length);
                    _.each(calendar.item, function(item, i) {
                        var guid = item.guid[0];
                        var id = parseInt(guid.split('/').pop(), 10);
                        assert.strictEqual(id, expectedEvents[i].id);
                        assert.strictEqual(item.title[0], expectedEvents[i].displayName);
                        assert.strictEqual(item['ev:organizer'][0], expectedEvents[i].organisers.sort().join());
                    });
                }
            }

            return callback(calendar);
        });
    });
};

/**
 * Assert that the RSS calendar for a serie can not be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie for wich to get the event calendar
 * @param  {String}             start                           The start date for the calendar
 * @param  {String}             end                             The end date for the calendar
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetSeriesCalendarRssFails = module.exports.assertGetSeriesCalendarRssFails = function(client, id, start, end, code, callback) {
    client.serie.getSeriesCalendarRss(id, start, end, function(err, calendar) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        assert.ok(!calendar);
        return callback();
    });
};

/**
 * Assert that the iCal event calendar for a serie can be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie for wich to get the event calendar
 * @param  {String}             [start]                         The start date for the calendar
 * @param  {String}             [end]                           The end date for the calendar
 * @param  {Event[]}            [expectedEvents]                The event(s) that should be in this calendar
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetSeriesCalendarIcal = module.exports.assertGetSeriesCalendarIcal = function(client, id, start, end, expectedEvents, callback) {
    client.serie.getSeriesCalendarIcal(id, start, end, function(err, calendar, response) {
        assert.ok(!err);

        // Assert the response is iCal
        assert.strictEqual(response.headers['content-type'], 'text/calendar; charset=utf-8');

        // Parse the iCal calendar
        TestsUtil.parseIcalCalendar(calendar, function(calendar) {

            // Check if the returned events are the ones we expected
            if (expectedEvents) {
                assert.strictEqual(calendar.subComponents.length, expectedEvents.length);
                _.each(calendar.subComponents, function(subComponent, i) {
                    var id = parseInt(subComponent.model.uid, 10);
                    assert.strictEqual(id, expectedEvents[i].id);
                    assert.strictEqual(subComponent.model.summary, expectedEvents[i].displayName);
                });
            }

            return callback(calendar);
        });
    });
};

/**
 * Assert that the iCal calendar for a serie can not be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie for wich to get the event calendar
 * @param  {String}             [start]                         The start date for the calendar
 * @param  {String}             [end]                           The end date for the calendar
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetSeriesCalendarIcalFails = module.exports.assertGetSeriesCalendarIcalFails = function(client, id, start, end, code, callback) {
    client.serie.getSeriesCalendarIcal(id, start, end, function(err, calendar) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        assert.ok(!calendar);
        return callback();
    });
};

/**
 * Assert that a user can be subscribed to a serie
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie that the user should be subscribed to
 * @param  {Number}             [userId]                        The id of the user that should be subscribed to the serie
 * @param  {Number}             [orgUnit]                       The id of the organisational unit that the serie belonged to when the user subscribed to it
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertSubscribeSeries = module.exports.assertSubscribeSeries = function(client, id, userId, orgUnit, callback) {
    client.serie.subscribeSeries(id, userId, orgUnit, function(err) {
        assert.ok(!err);

        // If no `userId` was specified, the current user would have been
        // subscribed. We get the current user's user id from the me feed
        UsersTestsUtil.assertGetMe(client, function(me) {
            if (!userId) {
                userId = me.id;
            }

            // Get the events in this serie so we can verify they appear in the user's calendar
            assertGetSeriesEvents(client, id, null, null, null, null, function(events) {

                // Generate a broad window for the user's calendar
                var start = moment().subtract(30, 'days').format();
                var end = moment().add(30, 'days').format();

                // If the serie contained events, we generate a window around the serie events
                // and assert that the events are returned in the user's calendar
                if (!_.isEmpty(events.results.length)) {
                    start = moment(new Date(_.first(events.results).start)).subtract(1, 'hours').format();
                    end = moment(new Date(_.last(events.results).end)).add(1, 'hours').format();
                }

                // Get the user's calendar
                UsersTestsUtil.assertGetUserCalendar(client, userId, start, end, events.results, function(calendar) {
                    return callback();
                });
            });
        });
    });
};

/**
 * Assert that a user can not be subscribed to a serie
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie that the user should be subscribed to
 * @param  {Number}             [userId]                        The id of the user that should be subscribed to the serie
 * @param  {Number}             [context]                       The id of the organisational unit that the serie belonged to when the user subscribed to it
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertSubscribeSeriesFails = module.exports.assertSubscribeSeriesFails = function(client, id, userId, context, code, callback) {
    client.serie.subscribeSeries(id, userId, context, function(err) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        return callback();
    });
};

/**
 * Assert that a user can be unsubscribed from a serie
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie that the user should be unsubscribed from
 * @param  {Number}             [userId]                        The id of the user that should be unsubscribed from the serie
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertUnsubscribeSeries = module.exports.assertUnsubscribeSeries = function(client, id, userId, callback) {
    client.serie.unsubscribeSeries(id, userId, function(err) {
        assert.ok(!err);

        // If no `userId` was specified, the current user would have been
        // unsubscribed. We get the current user's user id from the me feed
        UsersTestsUtil.assertGetMe(client, function(me) {
            if (!userId) {
                userId = me.id;
            }

            // Get the events in this serie so we can verify they no longer appear in the user's calendar
            assertGetSeriesEvents(client, id, null, null, null, null, function(serieEvents) {

                // Generate a broad window for the user's calendar
                var start = moment().subtract(30, 'days').format();
                var end = moment().add(30, 'days').format();

                // If the serie contained events, we generate a window around the events
                // and assert that the event does not return in the user's calendar
                if (!_.isEmpty(serieEvents.results.length)) {
                    start = moment(new Date(_.first(serieEvents.results).start)).subtract(1, 'hours').format();
                    end = moment(new Date(_.last(serieEvents.results).end)).add(1, 'hours').format();
                }

                // Get the user's calendar
                UsersTestsUtil.assertGetUserCalendar(client, userId, start, end, null, function(calendar) {
                    _.each(calendar.results, function(event) {
                        assert.ok(!_.find(serieEvents, {'id': event.id}));
                    });

                    return callback();
                });
            });
        });
    });
};

/**
 * Assert that a user can not be unsubscribed from a serie
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             id                              The id of the serie that the user should be unsubscribed from
 * @param  {Number}             userId                          The id of the user that should be unsubscribed from the serie
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertUnsubscribeSeriesFails = module.exports.assertUnsubscribeSeriesFails = function(client, id, userId, code, callback) {
    client.serie.unsubscribeSeries(id, userId, function(err) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        return callback();
    });
};

/**
 * Generate one or more series
 *
 * @param  {RestClient}         client                  The REST client to make the requests with
 * @param  {Number}             nrOfSeries              The number of series that should be created
 * @param  {Function}           callback                Standard callback function
 * @param  {Serie[]}            callback.series         The created series
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var generateSeries = module.exports.generateSeries = function(client, nrOfSeries, callback, _series) {
    _series = _series || [];
    if (nrOfSeries === 0) {
        return callback(_series);
    }

    // Create the serie
    var displayName = TestsUtil.generateString(30);
    assertCreateSerie(client, displayName, null, function(serie) {
        _series.push(serie);

        // Move on to the next serie
        generateSeries(client, nrOfSeries - 1, callback, _series);
    });
};

/**
 * Generate one or more series that each have a number of events
 *
 * @param  {RestClient}         client                  The REST client to make the requests with
 * @param  {Number}             nrOfSeries              The number of series that should be created
 * @param  {Number}             eventsPerSerie          The number of events in each serie
 * @param  {String}             rangeStart              The datetime when the events can start
 * @param  {String}             rangeEnd                The datetime when the events should end
 * @param  {Function}           callback                Standard callback function
 * @param  {Serie[]}            callback.series         The created series. Each serie will also hold an `events` key that lists all the events that were created for the serie
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var generateSerieWithEvents = module.exports.generateSerieWithEvents = function(client, nrOfSeries, eventsPerSerie, rangeStart, rangeEnd, callback, _series) {
    _series = _series || [];
    if (nrOfSeries === 0) {
        return callback(_series);
    }

    // Create the serie
    var displayName = TestsUtil.generateString(30);
    assertCreateSerie(client, displayName, null, function(serie) {

        // Create the events, if any
        EventsTestsUtil.generateTestEvents(client, eventsPerSerie, rangeStart, rangeEnd, function(events) {

            // Add the events to the serie
            var eventIds = _.pluck(events, 'id');
            assertAddSeriesEvents(client, serie.id, eventIds, function() {

                // Get the serie object as the metadata will have changed
                assertGetSeries(client, serie.id, false, true, null, function(serie) {
                    serie.events = events;
                    _series.push(serie);

                    // Move on to the next serie
                    generateSerieWithEvents(client, nrOfSeries - 1, eventsPerSerie, rangeStart, rangeEnd, callback, _series);
                });
            });
        });
    });
};

/**
 * Get the events for a set of series
 *
 * @param  {RestClient}     client                      The REST client to make the requests with
 * @param  {Number[]}       seriesIds                   The ids of the series to retrieve the events for
 * @param  {Function}       callback                    Standard callback function
 * @param  {Object}         callback.eventsPerSeries    The events keyed by the series id
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var getAllSeriesEvents = module.exports.getAllSeriesEvents = function(client, seriesIds, callback, _eventsPerSeries) {
    _eventsPerSeries = _eventsPerSeries || {};

    if (_.isEmpty(seriesIds)) {
        return callback(_eventsPerSeries);
    }

    var seriesId = seriesIds.pop();
    assertGetSeriesEvents(client, seriesId, 10000, 0, null, null, function(events) {
        _eventsPerSeries[seriesId] = events.results;
        getAllSeriesEvents(client, seriesIds, callback, _eventsPerSeries);
    });
};
