// gh-apps/tests/util.js
/**
 * Copyright (c) 2015 "Fronteer LTD"
 * Grasshopper Event Engine
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

var _ = require('lodash');
var assert = require('assert');

var ConfigAPI = require('gh-config');

/**
 * Assert that an app has all expected properties
 *
 * @param  {App}                app                     The app to assert the properties for
 * @param  {App}                expectedApp             The app to which the provided app should be compared
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var assertApp = module.exports.assertApp = function(app, expectedApp) {
    assert.ok(app);
    assert.ok(app.id);
    assert.ok(app.displayName);
    assert.ok(app.type);
    assert.ok(!_.isUndefined(app.enabled));
    assert.ok(app.createdAt);
    assert.ok(app.updatedAt);

    assert.strictEqual(app.id, expectedApp.id);
    assert.strictEqual(app.displayName, expectedApp.displayName);
    assert.strictEqual(app.type, expectedApp.type);
    assert.strictEqual(app.enabled, expectedApp.enabled);
    assert.strictEqual(app.createdAt, expectedApp.createdAt);
    assert.strictEqual(app.updatedAt, expectedApp.updatedAt);
};

/**
 * Assert that all apps for a tenant can be retrieved
 *
 * @param  {RestClient}         client                  The REST client to make the request with
 * @param  {Number}             tenantId                The id of the tenant to retrieve the apps for
 * @param  {Number}             [expectedTotal]         The expected total number of apps
 * @param  {Function}           callback                Standard callback function
 * @param  {Tenants[]}          callback.apps           All available apps in the tenant
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var assertGetApps = module.exports.assertGetApps = function(client, tenantId, expectedTotal, callback) {
    client.app.getApps(tenantId, function(err, apps) {
        assert.ok(!err);
        assert.ok(apps);

        if (expectedTotal) {
            assert.strictEqual(apps.length, expectedTotal);
        } else {
            assert.ok(apps.length >= 1);
        }

        _.each(apps, function(app) {
            assert.strictEqual(app.TenantId, tenantId);
        });

        return callback(apps);
    });
};

/**
 * Assert that all app for a tenant can not be retrieved
 *
 * @param  {RestClient}         client                  The REST client to make the request with
 * @param  {Number}             tenantId                The id of the tenant to retrieve the apps for
 * @param  {Number}             code                    The expected HTTP error code
 * @param  {Function}           callback                Standard callback function
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var assertGetAppsFails = module.exports.assertGetAppsFails = function(client, tenantId, code, callback) {
    client.app.getApps(tenantId, function(err, apps) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        assert.ok(!apps);

        return callback();
    });
};

/**
 * Assert that an app can be retrieved
 *
 * @param  {RestClient}         client                  The REST client to make the request with
 * @param  {Number}             id                      The id of the app to retrieve
 * @param  {App}                [expectedApp]           The expected app to be retrieved
 * @param  {Function}           callback                Standard callback function
 * @param  {App}                callback.app            The requested app
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var assertGetApp = module.exports.assertGetApp = function(client, id, expectedApp, callback) {
    client.app.getApp(id, function(err, app) {
        assert.ok(!err);
        assert.ok(app);
        assert.ok(app.id);
        assert.ok(app.TenantId);
        assert.ok(app.displayName);
        assert.ok(app.host);
        assert.ok(!_.isUndefined(app.enabled));
        assert.ok(app.type);
        assert.ok(app.createdAt);
        assert.ok(app.updatedAt);

        if (expectedApp) {
            assert.strictEqual(app.id, expectedApp.id);
            assert.strictEqual(app.TenantId, expectedApp.TenantId);
            assert.strictEqual(app.displayName, expectedApp.displayName);
            assert.strictEqual(app.host, expectedApp.host);
            assert.strictEqual(app.enabled, expectedApp.enabled);
            assert.strictEqual(app.type, expectedApp.type);
            assert.strictEqual(app.createdAt, expectedApp.createdAt);
            assert.strictEqual(app.updatedAt, expectedApp.updatedAt);
        }

        return callback(app);
    });
};

/**
 * Assert that an app can not be retrieved
 *
 * @param  {RestClient}         client                  The REST client to make the request with
 * @param  {Number}             id                      The id of the app to retrieve
 * @param  {Number}             code                    The expected HTTP error code
 * @param  {Function}           callback                Standard callback function
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var assertGetAppFails = module.exports.assertGetAppFails = function(client, id, code, callback) {
    client.app.getApp(id, function(err, app) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        assert.ok(!app);

        return callback();
    });
};

/**
 * Assert that the administrators for an app can be retrieved
 *
 * @param  {RestClient}         client                  The REST client to make the request with
 * @param  {Number}             id                      The id of the app to retrieve the administrators for
 * @param  {Number}             [limit]                 The maximum number of results to retrieve
 * @param  {Number}             [offset]                The paging number of the results to retrieve
 * @param  {Number}             [expectedResults]       The expected number of app administrators in the returned results
 * @param  {Number}             [expectedTotal]         The expected total number of app administrators
 * @param  {Function}           callback                Standard callback function
 * @param  {UserList}           callback.appAdmins      The administrators for the app
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var assertGetAppAdmins = module.exports.assertGetAppAdmins = function(client, id, limit, offset, expectedResults, expectedTotal, callback) {
    client.app.getAppAdmins(id, limit, offset, function(err, appAdmins) {
        assert.ok(!err);
        assert.ok(appAdmins);
        assert.ok(!_.isUndefined(appAdmins.count));
        assert.ok(!_.isUndefined(appAdmins.rows));

        if (expectedResults) {
            assert.strictEqual(appAdmins.rows.length, expectedResults);
        }

        if (expectedTotal) {
            assert.strictEqual(appAdmins.count, expectedTotal);
        }

        _.each(appAdmins.rows, function(appAdmin) {
            assert.strictEqual(appAdmin.AppId, id);
        });

        return callback(appAdmins);
    });
};

/**
 * Assert that the administrators for an app can not be retrieved
 *
 * @param  {RestClient}         client                  The REST client to make the request with
 * @param  {Number}             id                      The id of the app to retrieve the administrators for
 * @param  {Number}             [limit]                 The maximum number of results to retrieve
 * @param  {Number}             [offset]                The paging number of the results to retrieve
 * @param  {Number}             code                    The expected HTTP error code
 * @param  {Function}           callback                Standard callback function
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var assertGetAppAdminsFails = module.exports.assertGetAppAdminsFails = function(client, id, limit, offset, code, callback) {
    client.app.getAppAdmins(id, limit, offset, function(err, appAdmins) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        assert.ok(!appAdmins);

        return callback();
    });
};

/**
 * Assert that a new app can be created
 *
 * @param  {RestClient}         client                  The REST client to make the request with
 * @param  {Number}             tenantId                The id of the tenant on which to create the app
 * @param  {String}             displayName             The name of the app (e.g. `Timetable 2013-14`)
 * @param  {String}             host                    The host on which the app can be found
 * @param  {String}             type                    The type of the app
 * @param  {Function}           callback                Standard callback function
 * @param  {App}                callback.app            The created app
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var assertCreateApp = module.exports.assertCreateApp = function(client, tenantId, displayName, host, type, callback) {
    var createdApp = null;
    var isRequestDone = false;
    var isCached = false;

    /*!
     * This function will return the created application to the caller
     * when the REST request has finished and its config has been cached
     */
    var done = function() {
        if (isRequestDone && isCached) {
            return callback(createdApp);
        }
    };

    // Only return to the caller once the config has been cached
    ConfigAPI.once('cached', function() {
        isCached = true;
        done();
    });

    // Create the application
    client.app.createApp(tenantId, displayName, host, type, function(err, _createdApp) {
        assert.ok(!err);
        createdApp = _createdApp;
        assert.ok(createdApp);
        assert.ok(createdApp.id);
        assert.ok(createdApp.createdAt);
        assert.ok(createdApp.updatedAt);
        assert.strictEqual(createdApp.TenantId, tenantId);
        assert.strictEqual(createdApp.displayName, displayName);
        assert.strictEqual(createdApp.host, host);
        assert.strictEqual(createdApp.type, type);
        isRequestDone = true;
        done();
    });
};

/**
 * Assert that a new app can not be created
 *
 * @param  {RestClient}         client                  The REST client to make the request with
 * @param  {Number}             tenantId                The id of the tenant on which to create the app
 * @param  {String}             displayName             The name of the app (e.g. `Timetable 2013-14`)
 * @param  {String}             host                    The host on which the app can be found
 * @param  {String}             type                    The type of the app
 * @param  {Number}             code                    The expected HTTP error code
 * @param  {Function}           callback                Standard callback function
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var assertCreateAppFails = module.exports.assertCreateAppFails = function(client, tenantId, displayName, host, type, code, callback) {
    client.app.createApp(tenantId, displayName, host, type, function(err, createdApp) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        assert.ok(!createdApp);

        return callback();
    });
};

/**
 * Assert that an app can be updated
 *
 * @param  {RestClient}         client                  The REST client to make the request with
 * @param  {App}                app                     The app to update
 * @param  {Object}             opts                    The values to update the app with
 * @param  {String}             [opst.displayName]      Updated app name
 * @param  {Boolean}            [opts.enabled]          Whether the app should be enabled
 * @param  {String}             [opts.host]             Updated app host
 * @param  {Function}           callback                Standard callback function
 * @param  {App}                callback.app            The updated app
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var assertUpdateApp = module.exports.assertUpdateApp = function(client, app, opts, callback) {
    client.app.updateApp(app.id, opts, function(err, updatedApp) {
        assert.ok(!err);
        assert.ok(updatedApp);
        assert.strictEqual(updatedApp.id, app.id);
        assert.strictEqual(updatedApp.createdAt, app.createdAt);
        assert.notEqual(updatedApp.updatedAt, app.updatedAt);

        if (opts.displayName) {
            assert.strictEqual(updatedApp.displayName, opts.displayName);
        } else {
            assert.strictEqual(updatedApp.displayName, app.displayName);
        }

        if (!_.isUndefined(opts.enabled)) {
            assert.strictEqual(updatedApp.enabled, opts.enabled);
        } else {
            assert.strictEqual(updatedApp.enabled, app.enabled);
        }

        if (opts.host) {
            assert.strictEqual(updatedApp.host, opts.host);
        } else {
            assert.strictEqual(updatedApp.host, app.host);
        }

        return callback(updatedApp);
    });
};

/**
 * Assert that an app can not be updated
 *
 * @param  {RestClient}         client                  The REST client to make the request with
 * @param  {Number}             id                      The id of the app to update
 * @param  {Object}             opts                    The values to update
 * @param  {String}             [opts.displayName]      Updated app name
 * @param  {Boolean}            [opts.enabled]          Whether the app should be enabled
 * @param  {String}             [opts.host]             Updated app host
 * @param  {Number}             code                    The expected HTTP error code
 * @param  {Function}           callback                Standard callback function
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var assertUpdateAppFails = module.exports.assertUpdateAppFails = function(client, id, opts, code, callback) {
    client.app.updateApp(id, opts, function(err, updatedApp) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        assert.ok(!updatedApp);

        return callback();
    });
};

/**
 * Assert that the administrators for an app can be updated
 *
 * @param  {RestClient}         client                  The REST client to make the request with
 * @param  {Number}             id                      The id of the app for which to update the administrators
 * @param  {Object}             adminUpdates            Object that describes the app administrator changes to apply
 * @param  {Function}           callback                Standard callback function
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var assertUpdateAppAdmins = module.exports.assertUpdateAppAdmins = function(client, id, adminUpdates, callback) {
    client.app.updateAppAdmins(id, adminUpdates, function(err) {
        assert.ok(!err);
        return callback();
    });
};

/**
 * Assert that the administrators for an app can not be updated
 *
 * @param  {RestClient}         client                  The REST client to make the request with
 * @param  {Number}             id                      The id of the app for which to update the administrators
 * @param  {Object}             adminUpdates            Object that describes the app administrator changes to apply
 * @param  {Number}             code                    The expected HTTP error code
 * @param  {Function}           callback                Standard callback function
 * @throws {AssertionError}                             Error thrown when an assertion failed
 */
var assertUpdateAppAdminsFails = module.exports.assertUpdateAppAdminsFails = function(client, id, adminUpdates, code, callback) {
    client.app.updateAppAdmins(id, adminUpdates, function(err) {
        assert.ok(err);
        assert.strictEqual(err.code, code);

        return callback();
    });
};

/**
 * Assert that the Terms and Conditions for an application can be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             app                             The id of the app for which to get the Terms and Conditions
 * @param  {Function}           callback                        Standard callback function
 * @param  {Object}             callback.termsAndConditions     The Terms and Conditions
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetTermsAndConditions = module.exports.assertGetTermsAndConditions = function(client, app, callback) {
    client.app.getTermsAndConditions(app, function(err, termsAndConditions) {
        assert.ok(!err);
        assert.ok(_.isObject(termsAndConditions));
        assert.ok(_.isString(termsAndConditions.text));
        assert.ok(_.isString(termsAndConditions.lastUpdate));
        return callback(termsAndConditions);
    });
};

/**
 * Assert that the Terms and Conditions for an application can not be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Number}             app                             The id of the app for which to get the Terms and Conditions
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetTermsAndConditionsFails = module.exports.assertGetTermsAndConditionsFails = function(client, app, code, callback) {
    client.app.getTermsAndConditions(app, function(err, termsAndConditions) {
        assert.ok(err);
        assert.strictEqual(err.code, code);
        assert.ok(!termsAndConditions);
        return callback();
    });
};
