// gh-tenants/tests/util.js
/**
 * Copyright (c) 2014 "Fronteer LTD"
 * Grasshopper Event Engine
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

var assert = require('assert');

/**
 * TODO
 */
var assertTenant = module.exports.assertTenant = function(tenant, expectedTenant) {
    assert.ok(tenant);
    assert.ok(tenant.id);
    assert.ok(tenant.displayName);
    assert.ok(tenant.createdAt);
    assert.ok(tenant.updatedAt);
    assert.equal(tenant.id, expectedTenant.id);
    assert.equal(tenant.displayName, expectedTenant.displayName);
    assert.equal(tenant.createdAt, expectedTenant.createdAt);
    assert.equal(tenant.updatedAt, expectedTenant.updatedAt);
};

/**
 * TODO
 */
var assertGetTenants = module.exports.assertGetTenants = function(client, expectedTotal, callback) {
    client.tenant.getTenants(function(err, tenants) {
        assert.ok(!err);
        assert.ok(tenants);

        // TODO
        if (expectedTotal) {
            assert.equal(tenants.length, expectedTotal);
        } else {
            assert.ok(tenants.length > 1);
        }

        return callback(tenants);
    });
};

/**
 * TODO
 */
var assertGetTenantsFails = module.exports.assertGetTenantsFails = function(client, code, callback) {
    client.tenant.getTenants(function(err, tenants) {
        assert.ok(err);
        assert.equal(err.code, code);
        assert.ok(!tenants);

        return callback();
    });
};

/**
 * TODO
 */
var assertGetTenant = module.exports.assertGetTenant = function(client, id, expectedTenant, callback) {
    client.tenant.getTenant(id, function(err, tenant) {
        assert.ok(!err);
        assert.ok(tenant);
        assert.ok(tenant.id);
        assert.ok(tenant.displayName);
        assert.ok(tenant.createdAt);
        assert.ok(tenant.updatedAt);

        // TODO
        if (expectedTenant) {
            assert.equal(tenant.id, expectedTenant.id);
            assert.equal(tenant.displayName, expectedTenant.displayName);
            assert.equal(tenant.createdAt, expectedTenant.createdAt);
            assert.equal(tenant.updatedAt, expectedTenant.updatedAt);
        }

        return callback(tenant);
    });
};

/**
 * TODO
 */
var assertGetTenantFails = module.exports.assertGetTenantFails = function(client, id, code, callback) {
    client.tenant.getTenant(id, function(err, tenant) {
        if (!err) {
            console.log(tenant);
        }
        assert.ok(err);
        assert.equal(err.code, code);
        assert.ok(!tenant);

        return callback();
    });
};

/**
 * TODO
 */
var assertCreateTenant = module.exports.assertCreateTenant = function(client, displayName, callback) {
    client.tenant.createTenant(displayName, function(err, createdTenant) {
        assert.ok(!err);
        assert.ok(createdTenant);
        assert.ok(createdTenant.id);
        assert.ok(createdTenant.createdAt);
        assert.ok(createdTenant.updatedAt);
        assert.equal(createdTenant.displayName, displayName);

        return callback(createdTenant);
    });
};

/**
 * TODO
 */
var assertCreateTenantFails = module.exports.assertCreateTenantFails = function(client, displayName, code, callback) {
    client.tenant.createTenant(displayName, function(err, createdTenant) {
        assert.ok(err);
        assert.equal(err.code, code);
        assert.ok(!createdTenant);

        return callback();
    });
};

/**
 * TODO
 */
var assertUpdateTenant = module.exports.assertUpdateTenant = function(client, tenant, displayName, callback) {
    client.tenant.updateTenant(tenant.id, displayName, function(err, updatedTenant) {
        assert.ok(!err);
        assert.ok(updatedTenant);
        assert.equal(updatedTenant.id, tenant.id);
        assert.equal(updatedTenant.displayName, displayName);
        assert.equal(updatedTenant.createdAt, tenant.createdAt);
        assert.notEqual(updatedTenant.updatedAt, tenant.updatedAt);

        return callback(updatedTenant);
    });
};

/**
 * TODO
 */
var assertUpdateTenantFails = module.exports.assertUpdateTenantFails = function(client, id, displayName, code, callback) {
    client.tenant.updateTenant(id, displayName, function(err, updatedTenant) {
        assert.ok(err);
        assert.equal(err.code, code);
        assert.ok(!updatedTenant);

        return callback();
    });
};
