// gh-orgunit/lib/api.js
/**
 * Copyright (c) 2015 "Fronteer LTD"
 * Grasshopper Event Engine
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

var _ = require('lodash');
var util = require('util');

var AppsDAO = require('gh-apps/lib/internal/dao');
var CalendarInfo = require('gh-calendar/lib/model').CalendarInfo;
var CalendarUtil = require('gh-calendar/lib/util');
var GrasshopperUtil = require('gh-core/lib/util');
var GroupsAPI = require('gh-groups');
var GroupsDAO = require('gh-groups/lib/internal/dao');
var log = require('gh-core/lib/logger').logger('gh-orgunit');
var SeriesDAO = require('gh-series/lib/internal/dao');
var UsersDAO = require('gh-users/lib/internal/dao');
var Validator = require('gh-core/lib/validator').Validator;

var OrgUnitAuthz = require('./authz');
var OrgUnitDAO = require('./internal/dao');

/**
 * Create an organisational unit
 *
 * @param  {Context}        ctx                 Standard context containing the current user and the current app
 * @param  {Number}         appId               The id of the application on which to create the organisational unit
 * @param  {String}         displayName         The name of the organisational unit
 * @param  {String}         type                The organisational unit type (e.g. `tripos`, `part`)
 * @param  {String}         [description]       The description of the organisational unit
 * @param  {Object|String}  [metadata]          The extra metadata for the organisational unit. When using a string, it should be a stringified JSON object
 * @param  {Boolean}        [published]         The flag that identifies the published status of the organisational unit
 * @param  {Number}         [groupId]           The id of the group that can manage the organisational unit. Defaults to creating a new group with the current user as a member
 * @param  {Number}         [parentId]          The id of the parent organisational unit. By not providing a parent id, you will end up with a top level organisational unit
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error object, if any
 * @param  {OrgUnit}        callback.orgUnit    The created organisational unit
 */
var createOrgUnit = module.exports.createOrgUnit = function(ctx, appId, displayName, type, description, metadata, published, groupId, parentId, callback) {
    // Ensure that the app id is a valid number
    appId = GrasshopperUtil.getNumberParam(appId);
    metadata = metadata || {};

    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'You need to be authenticated in order to create an organisational unit'}).isLoggedInUser(ctx);
    validator.check(appId, {'code': 400, 'msg': 'A valid application id must be provided'}).isInt();
    validator.check(displayName, {'code': 400, 'msg': 'A valid displayName must be provided'}).notEmpty();
    validator.check(displayName, {'code': 400, 'msg': 'A displayName can be at most 255 characters long'}).isShortString();
    validator.check(type, {'code': 400, 'msg': 'A valid type must be provided'}).notEmpty();
    validator.check(type, {'code': 400, 'msg': 'A type can be at most 32 characters long'}).len(1, 32);
    if (description) {
        validator.check(description, {'code': 400, 'msg': 'A description can be at most 1000 characters long'}).isMediumString();
    }
    if (groupId) {
        validator.check(groupId, {'code': 400, 'msg': 'A valid group id must be provided'}).isInt();
        groupId = GrasshopperUtil.getNumberParam(groupId);
    }
    if (parentId) {
        validator.check(parentId, {'code': 400, 'msg': 'A valid parent organisational unit id must be provided'}).isInt();
        parentId = GrasshopperUtil.getNumberParam(parentId);
    }
    if (metadata) {
        if (_.isString(metadata)) {
            try {
                metadata = JSON.parse(metadata);
            } catch (e) {
                var actor = (ctx.user ? ctx.user.id : 'anonymous');
                log().warn({'actor': actor, 'displayName': displayName}, 'Invalid JSON was submitted');
            }
        }
        validator.check(null, {'code': 400, 'msg': 'A metadata object must be provided in JSON'}).isObject(metadata);
    }
    if (published) {
        published = GrasshopperUtil.getBooleanParam(published);
        validator.check(null, {'code': 400, 'msg': 'A valid published flag must be provided'}).isBoolean(published);
    }
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // If a parent was specified, ensure it exists
    _getOptionalOrgUnit(ctx, parentId, function(err, parent) {
        if (err) {
            return callback(err);

        // Ensure the parent is from the same application
        } else if (parent && parent.AppId !== appId) {
            return callback({'code': 400, 'msg': 'You cannot create an organisational unit under an organisational unit from another application'});
        }

        // Ensure the current user can create an organisational unit
        OrgUnitAuthz.canCreateOrgUnit(ctx, appId, parent, function(err, canCreate) {
            if (err) {
                return callback(err);
            } else if (!canCreate) {
                log().warn({
                    'actor': ctx.user.id,
                    'displayName': displayName,
                    'parentId': parentId
                }, 'An unauthorized attempt at creating an organisational unit occurred');
                return callback({'code': 401, 'msg': 'You are not allowed to create an organisational unit'});
            }

            // Ensure the application exists
            AppsDAO.getApp(appId, function(err, app) {
                if (err) {
                    return callback(err);
                }

                // If a group was specified, ensure it exists. Create
                // a group if none was specified
                GroupsAPI.getOrCreateGroup(ctx, appId, groupId, function(err, group) {
                    if (err) {
                        return callback(err);
                    }

                    // Create the organisational unit
                    var opts = {
                        'description': description,
                        'metadata': metadata,
                        'published': published,
                        'parentId': parentId
                    };
                    return OrgUnitDAO.createOrgUnit(appId, displayName, type, group.id, opts, callback);
                });
            });
        });
    });
};

/**
 * Get an organisational unit
 *
 * @param  {Context}        ctx                 Standard context containing the current user and the current app
 * @param  {Number}         id                  The id of the organisational unit to retrieve
 * @param  {Boolean}        [includeSeries]     Whether to include the event series associated to the organisational units. Defaults to `false`
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error object, if any
 * @param  {OrgUnit}        callback.orgUnit    The retrieved organisational unit
 */
var getOrgUnit = module.exports.getOrgUnit = function(ctx, id, includeSeries, callback) {
    // Ensure that the id is a valid number
    id = GrasshopperUtil.getNumberParam(id);

    // Ensure that the includeSeries flag is a boolean
    includeSeries = GrasshopperUtil.getBooleanParam(includeSeries, false);

    var validator = new Validator();
    validator.check(id, {'code': 400, 'msg': 'A valid organisational unit id must be provided'}).isInt();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Ensure that the organisational unit exists
    OrgUnitDAO.getOrgUnit(id, includeSeries, function(err, orgUnit) {
        if (err) {
            return callback(err);
        } else if (orgUnit.AppId !== ctx.app.id && (!ctx.user || !ctx.user.canAdmin(orgUnit.AppId))) {
            return callback({'code': 401, 'msg': 'You cannot get an organisational unit from another application'});
        }

        return callback(null, orgUnit);
    });
};

/**
 * Get the organisational units for an application
 *
 * @param  {Context}        ctx                     Standard context containing the current user and the current app
 * @param  {Number}         appId                   The id of the app to get the organisational units for
 * @param  {Number}         [parentId]              The id of the parent to retrieve the organisational units for
 * @param  {String[]}       [types]                 The organisational unit type(s) to filter the organisational unit by
 * @param  {Boolean}        [includeSeries]         Whether to include the event series associated to the organisational units. Defaults to `false`
 * @param  {Boolean}        [includePermissions]    Whether to include if the current user can manage the organisational units/series and whether an organisational unit is locked. Defaults to `false`
 * @param  {Function}       callback                Standard callback function
 * @param  {Object}         callback.err            An error object, if any
 * @param  {OrgUnit[]}      callback.orgUnits       The matching organisational units
 */
var getOrgUnits = module.exports.getOrgUnits = function(ctx, appId, parentId, types, includeSeries, includePermissions, callback) {
    // Ensure that the app id is a valid number
    appId = GrasshopperUtil.getNumberParam(appId);

    // Ensure that the parent id (if any) is a valid number
    parentId = GrasshopperUtil.getNumberParam(parentId);

    // Ensure that the includeSeries flag is a valid boolean
    includeSeries = GrasshopperUtil.getBooleanParam(includeSeries);

    // Ensure that the includePermissions flag is a valid boolean
    includePermissions = GrasshopperUtil.getBooleanParam(includePermissions);

    var validator = new Validator();
    validator.check(appId, {'code': 400, 'msg': 'A valid application id must be provided'}).isInt();
    if (parentId) {
        validator.check(parentId, {'code': 400, 'msg': 'A valid organisational unit id must be provided'}).isInt();
    }
    if (types) {
        validator.check(null, {'code': 400, 'msg': 'Types must be specified as an array'}).isArray(types);
        _.each(types, function(type) {
            validator.check(type, {'code': 400, 'msg': 'A valid type must be between 1 and 32 characters long'}).len(1, 32);
        });
    }
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    } else if (appId !== ctx.app.id && (!ctx.user || !ctx.user.isGlobalAdmin())) {
        return callback({'code': 401, 'msg': 'You cannot get the organisational units from another application'});
    }

    // Get the parent organisational unit, if any
    _getOptionalOrgUnit(ctx, parentId, function(err, parent) {
        if (err) {
            return callback(err);
        } else if (parent && parent.AppId !== appId && !ctx.user.isGlobalAdmin()) {
            return callback({'code': 401, 'msg': 'You cannot get the organisational units under an organisational unit from another application'});
        }

        // Get the organisational units in the application
        var opts = {
            'includeSeries': includeSeries,
            'includeGroups': includePermissions
        };
        // If the current user is authenticated and we need to include the series,
        // we also include whether the user is subscribed to each serie
        if (ctx.user && includeSeries) {
            opts.subscriptions = {
                'include': true,
                'userId': ctx.user.id
            };
        }
        OrgUnitDAO.getOrgUnits(appId, parentId, types, opts, function(err, orgUnits) {
            if (err) {
                return callback(err);
            }

            // If we don't need to retrieve whether the current user can manage
            // the organisational units and series, we can return early
            if (!includePermissions || !ctx.user) {
                var augmentedOrgUnits = _augmentOrgUnits(ctx, orgUnits, []);
                return callback(null, augmentedOrgUnits);
            }

            // Otherwise we get the groups the current user is a member of so we can mark
            // those organisational units that the user can edit
            GroupsDAO.getUserGroups(ctx.user.id, 10000, 0, function(err, groups) {
                if (err) {
                    return callback(err);
                }

                var augmentedOrgUnits = _augmentOrgUnits(ctx, orgUnits, groups);
                return callback(null, augmentedOrgUnits);
            });
        });
    });
};

/**
 * Augment a set of organisational units with extra information. Each organisational
 * unit will be augmented with a `canManage` flag expressing whether the current
 * user can manage the unit. Each serie in the organisational unit will also be augmented
 * with an `isSubscribed` flag indicating whether the current user is subscribed to the serie
 *
 * @param  {Context}        ctx         Standard context containing the current user and the current app
 * @param  {OrgUnit[]}      orgUnits    The organisational units to augment
 * @param  {Group[]}        groups      The groups the current user is a member of
 * @return {OrgUnit[]}                  The augmented organisational units
 * @api private
 */
var _augmentOrgUnits = function(ctx, orgUnits, groups) {
    return _.map(orgUnits, function(orgUnit) {
        // Serialize the organisational unit
        var augmentedOrgUnit = orgUnit.toJSON();

        // Check if the current user is a member of the group that
        // this organisational unit belongs to
        var canManageThroughGroups = _contains(groups, orgUnit.GroupId);
        var canAdminOrgUnit = (ctx.user && ctx.user.canAdmin(orgUnit.AppId));
        augmentedOrgUnit.canManage = canManageThroughGroups || canAdminOrgUnit || false;

        // Add an `isSubscribed` flag to each serie
        augmentedOrgUnit.Series = _.map(orgUnit.Series, function(serie) {
            // Serialize the serie
            var augmentedSerie = serie.toJSON();

            // Determine whether the user is subscribed to the serie
            augmentedSerie.subscribed = !_.isEmpty(serie.Calendars);

            // Determine whether the user can manage the serie
            var canManageSerieThroughGroups = _contains(groups, serie.GroupId);
            var canAdminSerie = (ctx.user && ctx.user.canAdmin(serie.AppId));
            augmentedSerie.canManage = canManageSerieThroughGroups || canAdminSerie || false;

            // Sort the organisational units this series is under based on when they were added to the organisational units.
            // This allows us to check whether the series is borrowed from another organisational unit than the one it's
            // currently under. If this series is not borrowed we won't add anything in the output
            serie.OrgUnits = _.sortBy(serie.OrgUnits, function(orgUnit) {
                return orgUnit.OrgUnitsSeries.createdAt.getTime();
            });
            if (serie.OrgUnits[0].id !== orgUnit.id) {
                augmentedSerie.borrowedFrom = serie.OrgUnits[0];
            }

            // In order to save on serialization costs and bandwidth traffic we don't output the data we don't need
            delete augmentedSerie.OrgUnits;
            delete augmentedSerie.Calendars;

            return augmentedSerie;
        });

        return augmentedOrgUnit;
    });
};

/**
 * Check wheter a set of groups contains a group with a given id
 *
 * @param  {Group[]}    groups      The set of groups to look for a group with id `groupId` in
 * @param  {Number}     groupId     The id of the group to look for
 * @return {Boolean}                `true` if the set of groups contains the group
 * @api private
 */
var _contains = function(groups, groupId) {
    var group = _.find(groups, function(group) {
        return (group.id === groupId);
    });
    return _.isObject(group);
};

/**
 * Update an organisational unit
 *
 * @param  {Context}        ctx                     Standard context containing the current user and the current app
 * @param  {Number}         id                      The id of the organisational unit to update
 * @param  {Object}         opts                    The values to update
 * @param  {String}         [opts.displayName]      Updated organisational unit name
 * @param  {String}         [opts.description]      Updated organisational unit description
 * @param  {String}         [opts.type]             Updated organisational unit type
 * @param  {Object|String}  [opts.metadata]         Updated organisational unit metadata. When using a string, it should be a stringified JSON object
 * @param  {Boolean}        [opts.published]        Updated organisational unit published flag
 * @param  {Number}         [opts.groupId]          Updated organisational unit group
 * @param  {Number}         [opts.parent]           Updated organisational unit parent
 * @param  {Function}       callback                Standard callback function
 * @param  {Object}         callback.err            An error object, if any
 * @param  {OrgUnit}        callback.orgUnit        The updated organisational unit
 */
var updateOrgUnit = module.exports.updateOrgUnit = function(ctx, id, opts, callback) {
    // Ensure that the orgunit id is a valid number
    id = GrasshopperUtil.getNumberParam(id);

    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'You need to be authenticated in order to update an organisational unit'}).isLoggedInUser(ctx);
    validator.check(id, {'code': 400, 'msg': 'A valid orgunit id must be provided'}).isInt();
    validator.check(_.keys(opts).length, {'code': 400, 'msg': 'At least 1 update value must be provided'}).min(1);

    var validKeys = ['displayName', 'type', 'description', 'group', 'parent', 'metadata', 'published'];
    _.each(opts, function(value, key) {
        validator.check(key, {'code': 400, 'msg': 'An update must be one of: ' + validKeys.join(', ')}).isIn(validKeys);
    });

    var update = {};

    if (opts) {
        if (opts.displayName) {
            validator.check(opts.displayName, {'code': 400, 'msg': 'A valid displayName must be provided'}).notEmpty();
            validator.check(opts.displayName, {'code': 400, 'msg': 'A displayName can be at most 255 characters long'}).isShortString();
            update.displayName = opts.displayName;
        }
        if (opts.type) {
            validator.check(opts.type, {'code': 400, 'msg': 'A valid type must be provided'}).notEmpty();
            validator.check(opts.type, {'code': 400, 'msg': 'A type can be at most 32 characters long'}).len(1, 32);
            update.type = opts.type;
        }
        if (opts.description) {
            validator.check(opts.description, {'code': 400, 'msg': 'A description can be at most 1000 characters long'}).isMediumString();
            update.description = opts.description;
        }
        if (opts.group) {
            validator.check(opts.group, {'code': 400, 'msg': 'A valid group id must be provided'}).isInt();
            update.GroupId = GrasshopperUtil.getNumberParam(opts.group);
        }
        if (opts.parent) {
            validator.check(opts.parent, {'code': 400, 'msg': 'A valid parent organisational unit id must be provided'}).isInt();
            update.ParentId = GrasshopperUtil.getNumberParam(opts.parent);
        }
        if (!_.isUndefined(opts.published)) {
            opts.published = GrasshopperUtil.getBooleanParam(opts.published);
            validator.check(null, {'code': 400, 'msg': 'A valid published flag must be provided'}).isBoolean(opts.published);
            update.published = opts.published;
        }
        if (opts.metadata) {
            if (_.isString(opts.metadata)) {
                try {
                    opts.metadata = JSON.parse(opts.metadata);
                } catch (e) {
                    var actor = (ctx.user ? ctx.user.id : 'anonymous');
                    log().warn({'actor': actor, 'id': id}, 'Invalid JSON was submitted');
                }
            }
            validator.check(null, {'code': 400, 'msg': 'A valid metadata object must be provided'}).isObject(opts.metadata);
            update.metadata = opts.metadata;
        }
    }
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Check if the org unit exists
    OrgUnitDAO.getOrgUnit(id, false, function(err, orgUnit) {
        if (err) {
            return callback(err);
        }

        // Ensure the current user can update the organisational unit
        OrgUnitAuthz.canUpdateOrgUnit(ctx, orgUnit, function(err, canUpdate) {
            if (err) {
                return callback(err);
            } else if (!canUpdate) {
                log().warn({
                    'actor': ctx.user.id,
                    'id': orgUnit.id
                }, 'An unauthorized attempt at updating an organisational unit occurred');
                return callback({'code': 401, 'msg': 'You are not allowed to update the organisational unit'});
            }

            // If a parent was specified, ensure it exists
            _getOptionalOrgUnit(ctx, update.ParentId, function(err, parent) {
                if (err) {
                    return callback(err);

                // Ensure the parent is from the same application
                } else if (parent && parent.AppId !== orgUnit.AppId) {
                    return callback({'code': 400, 'msg': 'You cannot make an organisational unit from another application a parent of this one'});
                }

                // If a group was specified, ensure it exists
                _getOptionalGroup(ctx, update.GroupId, function(err, group) {
                    if (err) {
                        return callback(err);
                    } else if (group && group.AppId !== orgUnit.AppId) {
                        return callback({'code': 400, 'msg': 'You cannot use a group from another application'});
                    }

                    // Update the organisational unit
                    OrgUnitDAO.updateOrgUnit(orgUnit, update, callback);
                });
            });
        });
    });
};

/**
 * Delete an organisational unit
 *
 * @param  {Context}        ctx             Standard context containing the current user and the current app
 * @param  {Number}         id              The id of the organisational unit to delete
 * @param  {Function}       callback        Standard callback function
 * @param  {Object}         callback.err    An error object, if any
 */
var deleteOrgUnit = module.exports.deleteOrgUnit = function(ctx, id, callback) {
    // Ensure that the orgunit id is a valid number
    id = GrasshopperUtil.getNumberParam(id);

    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'You need to be authenticated in order to delete an organisational unit'}).isLoggedInUser(ctx);
    validator.check(id, {'code': 400, 'msg': 'A valid orgunit id must be provided'}).isInt();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Ensure the organisational unit exists
    OrgUnitDAO.getOrgUnit(id, false, function(err, orgUnit) {
        if (err) {
            return callback(err);
        }

        // Ensure the current user can delete the organisational unit
        OrgUnitAuthz.canDeleteOrgUnit(ctx, orgUnit, function(err, canDelete) {
            if (err) {
                return callback(err);
            } else if (!canDelete) {
                log().warn({
                    'actor': ctx.user.id,
                    'id': orgUnit.id
                }, 'An unauthorized attempt at deleting an organisational unit occurred');
                return callback({'code': 401, 'msg': 'You are not allowed to delete the organisational unit'});
            }

            // Delete the organisational unit
            OrgUnitDAO.deleteOrgUnit(orgUnit, callback);
        });
    });
};

/* Series */

/**
 * Get the event series in an organisational unit
 *
 * @param  {Context}        ctx                 Standard context containing the current user and the current app
 * @param  {number}         id                  The id of the organisational unit to retrieve the event series for
 * @param  {Number}         [limit]             The maximum number of results to retrieve. Default: 10
 * @param  {Number}         [offset]            The paging number of the results to retrieve
 * @param  {Boolean}        [upcoming]          Whether to only include event series with upcoming events. Defaults to `false`
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error object, if any
 * @param  {Serie[]}        callback.series     The retrieved event series
 */
var getOrgUnitSeries = module.exports.getOrgUnitSeries = function(ctx, id, limit, offset, upcoming, callback) {
    // Ensure that the paging values are valid
    limit = GrasshopperUtil.getNumberParam(limit, 10, 1, 25);
    offset = GrasshopperUtil.getNumberParam(offset, 0, 0);

    // Ensure that the upcoming flag is a boolean
    upcoming = GrasshopperUtil.getBooleanParam(upcoming, false);

    var validator = new Validator();
    validator.check(id, {'code': 400, 'msg': 'A valid organisational unit id must be provided'}).isInt();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Ensure the organisational unit exists
    getOrgUnit(ctx, id, false, function(err, orgUnit) {
        if (err) {
            return callback(err);
        }

        // Get the series for the organisational unit
        OrgUnitDAO.getOrgUnitSeries(orgUnit, limit, offset, upcoming, callback);
    });
};

/**
 * Add an event series to an organisational unit
 *
 * @param  {Context}        ctx                 Standard context containing the current user and the current app
 * @param  {Number}         id                  The id of the organisational unit to add the serie(s) to
 * @param  {Number[]}       series              The id(s) of the serie(s) to add to the organisational unit
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error object, if any
 */
var addOrgUnitSeries = module.exports.addOrgUnitSeries = function(ctx, id, series, callback) {
    // Ensure that the orgunit id is a valid number
    id = GrasshopperUtil.getNumberParam(id);

    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'You need to be authenticated in order to add series to an organisational unit'}).isLoggedInUser(ctx);
    validator.check(id, {'code': 400, 'msg': 'A valid organisational unit id must be provided'}).isInt();
    validator.check(null, {'code': 400, 'msg': 'At least 1 serie needs to be specified'}).isArray(series);
    if (series) {
        validator.check(series.length, {'code': 400, 'msg': 'At least 1 serie needs to be specified'}).min(1);
        _.each(series, function(serie) {
            validator.check(serie, {'code': 400, 'msg': 'A valid serie id must be provided'}).isInt();
        });
    }
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Ensure the organisational unit exists
    OrgUnitDAO.getOrgUnit(id, true, function(err, orgUnit) {
        if (err) {
            return callback(err);
        }

        // Ensure the series exist
        SeriesDAO.getSeries(series, function(err, series) {
            if (err) {
                return callback(err);
            }

            // Ensure the series are from the same application as the organisational unit
            var otherAppSeries = _.filter(series, function(serie) {
                return (orgUnit.AppId !== serie.AppId);
            });
            if (!_.isEmpty(otherAppSeries)) {
                return callback({'code': 400, 'msg': 'Some of the provided series belong to another application'});
            }

            // Get all the series that are NOT part of the organisational unit already
            var newSeries = _.filter(series, function(serie) {
                return !_.find(orgUnit.Series, function(orgUnitSerie) {
                    return (orgUnitSerie.id === serie.id);
                });
            });

            // If there are no new series, we can return early
            if (_.isEmpty(newSeries)) {
                return callback();
            }

            // Ensure the current user can add series to the organisational unit
            OrgUnitAuthz.canAddOrgUnitSeries(ctx, orgUnit, function(err, canAdd) {
                if (err) {
                    return callback(err);
                } else if (!canAdd) {
                    log().warn({
                        'actor': ctx.user.id,
                        'displayName': orgUnit.displayName
                    }, 'An unauthorized attempt at adding series to an organisational unit occurred');
                    return callback({'code': 401, 'msg': 'You are not allowed to add series to the organisational unit'});
                }

                OrgUnitDAO.addOrgUnitSeries(orgUnit, series, callback);
            });
        });
    });
};

/**
 * Remove an event series from an organisational unit
 *
 * @param  {Context}        ctx                 Standard context containing the current user and the current app
 * @param  {Number}         id                  The id of the organisational unit to remove the serie(s) from
 * @param  {Number[]}       series              The id(s) of the serie(s) to remove from the organisational unit
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error object, if any
 */
var deleteOrgUnitSeries = module.exports.deleteOrgUnitSeries = function(ctx, id, series, callback) {
    // Ensure that the orgunit id is a valid number
    id = GrasshopperUtil.getNumberParam(id);

    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'You need to be authenticated in order to delete series from an organisational unit'}).isLoggedInUser(ctx);
    validator.check(id, {'code': 400, 'msg': 'A valid organisational unit id must be provided'}).isInt();
    validator.check(null, {'code': 400, 'msg': 'At least 1 serie needs to be specified'}).isArray(series);
    if (series) {
        validator.check(series.length, {'code': 400, 'msg': 'At least 1 serie needs to be specified'}).min(1);
        _.each(series, function(serie) {
            validator.check(serie, {'code': 400, 'msg': 'A valid serie id must be provided'}).isInt();
        });
    }
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    series = _.map(series, function(serieId) {
        return GrasshopperUtil.getNumberParam(serieId);
    });

    // Ensure the organisational unit exists
    OrgUnitDAO.getOrgUnit(id, true, function(err, orgUnit) {
        if (err) {
            return callback(err);
        }

        // Ensure the series are part of the organisational unit
        var serieObjects = _.filter(orgUnit.Series, function(serie) {
            return _.contains(series, serie.id);
        });
        if (serieObjects.length !== series.length) {
            return callback({'code': 400, 'msg': 'One or more of the specified series are not part of the organisational unit'});
        }

        // Ensure the current user can delete series from the organisational unit
        OrgUnitAuthz.canDeleteOrgUnitSeries(ctx, orgUnit, function(err, canDelete) {
            if (err) {
                return callback(err);
            } else if (!canDelete) {
                log().warn({
                    'actor': ctx.user.id,
                    'displayName': orgUnit.displayName
                }, 'An unauthorized attempt at deleting series from an organisational unit occurred');
                return callback({'code': 401, 'msg': 'You are not allowed to delete series from the organisational unit'});
            }

            OrgUnitDAO.deleteOrgUnitSeries(orgUnit, serieObjects, callback);
        });
    });
};

/* Calendar */

/**
 * Get the calendar for an organisational unit
 *
 * @param  {Context}        ctx                 Standard context containing the current user and the current app
 * @param  {Number}         id                  The id of the organisational unit to get the calendar for
 * @param  {String}         start               The timestamp (ISO 8601) from which to get the calendar for the organisation unit
 * @param  {String}         end                 The timestamp (ISO 8601) until which to get the calendar for the organisation unit
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error object, if any
 * @param  {Event[]}        callback.events     The requested organisational unit calendar
 */
var getOrgUnitCalendar = module.exports.getOrgUnitCalendar = function(ctx, id, start, end, callback) {
    // Ensure that the organisational unit id is a valid number
    id = GrasshopperUtil.getNumberParam(id);

    var validator = new Validator();
    validator.check(id, {'code': 400, 'msg': 'A valid organisational unit id must be provided'}).isInt();
    validator.check(start, {'code': 400, 'msg': 'A start time must be provided in a valid date format'}).isDate();
    validator.check(end, {'code': 400, 'msg': 'An end time must be provided in a valid date format'}).isDate();
    validator.check(start, {'code': 400, 'msg': 'The start time must be before the end time'}).isBefore(end);
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Get the organisational unit's calendar
    _getOrgUnitCalendar(ctx, id, start, end, callback);
};

/**
 * Get the calendar for an organisational unit in iCal format
 *
 * @param  {Context}        ctx                 Standard context containing the current user and the current app
 * @param  {Number}         id                  The id of the organisational unit to get the calendar for in iCal format
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error object, if any
 * @param  {String}         callback.ical       The requested organisational unit calendar in iCal format
 */
var getOrgUnitCalendarIcal = module.exports.getOrgUnitCalendarIcal = function(ctx, id, callback) {
    // Ensure that the organisational unit id is a valid number
    id = GrasshopperUtil.getNumberParam(id);

    var validator = new Validator();
    validator.check(id, {'code': 400, 'msg': 'A valid organisational unit id must be provided'}).isInt();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Get the organisational unit's calendar
    _getOrgUnitCalendar(ctx, id, null, null, function(err, events, orgUnit) {
        if (err) {
            return callback(err);
        }

        // Return the events as iCal
        var calendarInfo = _getCalendarInfo(ctx, orgUnit, 'ical');
        var ical = CalendarUtil.eventsToICal(calendarInfo, events);
        return callback(null, ical);
    });
};

/**
 * Get the calendar for an organisational unit in RSS format
 *
 * @param  {Context}        ctx                 Standard context containing the current user and the current app
 * @param  {Number}         id                  The id of the organisational unit to get the calendar for in RSS format
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error object, if any
 * @param  {String}         callback.rss        The requested organisational unit calendar in RSS format
 */
var getOrgUnitCalendarRss = module.exports.getOrgUnitCalendarRss = function(ctx, id, callback) {
    // Ensure that the organisational unit id is a valid number
    id = GrasshopperUtil.getNumberParam(id);

    var validator = new Validator();
    validator.check(id, {'code': 400, 'msg': 'A valid organisational unit id must be provided'}).isInt();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Get the organisational unit's calendar
    _getOrgUnitCalendar(ctx, id, null, null, function(err, events, orgUnit) {
        if (err) {
            return callback(err);
        }

        // Return the events as RSS
        var calendarInfo = _getCalendarInfo(ctx, orgUnit, 'rss');
        var rss = CalendarUtil.eventsToRSS(calendarInfo, events);
        return callback(null, rss);
    });
};

/**
 * Get the calendar for an organisational unit
 *
 * @param  {Context}        ctx                 Standard context containing the current user and the current app
 * @param  {Number}         id                  The id of the organisational unit to get the calendar for
 * @param  {String}         [start]             The timestamp (ISO 8601) from which to get the calendar for the organisation unit
 * @param  {String}         [end]               The timestamp (ISO 8601) until which to get the calendar for the organisation unit
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.end        An error object, if any
 * @param  {Event[]}        callback.events     The requested organisational unit calendar
 * @param  {OrgUnit}        callback.orgUnit    The organisational unit
 * @api private
 */
var _getOrgUnitCalendar = function(ctx, id, start, end, callback) {
    getOrgUnit(ctx, id, false, function(err, orgUnit) {
        if (err) {
            return callback(err);
        }

        OrgUnitDAO.getOrgUnitCalendar(orgUnit, start, end, function(err, events) {
            if (err) {
                return callback(err);
            }

            return callback(err, events, orgUnit);
        });
    });
};

/**
 * Get the calendar info for an organisational unit
 *
 * @param  {Context}        ctx             Standard context containing the current user and the current app
 * @param  {OrgUnit}        orgUnit         The organisational unit for which to get the calendar info
 * @param  {String}         format          The format in which the calendar will be returned. One of `ical` or `rss`
 * @return {CalendarInfo}                   An object representing the calendar info for the organisational unit
 * @api private
 */
var _getCalendarInfo = function(ctx, orgUnit, format) {
    var link = util.format('https://%s/api/orgunit/%s/calendar.%s', ctx.app.host, orgUnit.id, format);
    return new CalendarInfo(ctx.app, orgUnit.displayName, '', ctx.app.displayName, link, '', orgUnit.updatedAt);
};

/* Utilities */

/**
 * Get an organisational unit. If no `orgUnitId` was specified, `null`
 * will be returned rather than a 404 error. If an id was specified
 * that does not point to an existing orgUnit a 404 error will be returned.
 *
 * @param  {Context}        ctx                     Standard context containing the current user and the current app
 * @param  {Number}         orgUnitId               The id of the organisational unit to retrieve
 * @param  {Function}       callback                Standard callback function
 * @param  {Object}         callback.err            An error object, if any
 * @param  {OrgUnit}        [callback.orgUnit]      The retrieved organisational unit or null, if the provided `orgUnitId` was `null`
 * @api private
 */
var _getOptionalOrgUnit = function(ctx, orgUnitId, callback) {
    if (!orgUnitId) {
        return callback(null, null);
    }

    OrgUnitDAO.getOrgUnit(orgUnitId, false, callback);
};

/**
 * Get a group. If no `groupId` was specified, `null` will be returned
 * rather than a 404 error. If an id was specified that does not point
 * to an existing orgUnit a 404 error will be returned.
 *
 * @param  {Context}        ctx                     Standard context containing the current user and the current app
 * @param  {Number}         groupId                 The id of the group to retrieve
 * @param  {Function}       callback                Standard callback function
 * @param  {Object}         callback.err            An error object, if any
 * @param  {Group}          [callback.group]        The retrieved group or null, if the provided `groupId` was `null`
 * @api private
 */
var _getOptionalGroup = function(ctx, groupId, callback) {
    if (!groupId) {
        return callback(null, null);
    }

    GroupsDAO.getGroup(groupId, callback);
};

/* Subscriptions */

/**
 * Subscribe to the event series and events in an organisational unit
 *
 * @param  {Context}        ctx                 Standard context containing the current user and the current app
 * @param  {Number}         id                  The id of the organisational unit to subscribe to the event series and events for
 * @param  {Number}         [userId]            The id of the user that should be subscribed. Defaults to the current user
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error object, if any
 */
var subscribeOrgUnit = module.exports.subscribeOrgUnit = function(ctx, id, userId, callback) {
    // Default to the current user if no user was specified
    if (!userId && ctx.user) {
        userId = ctx.user.id;
    }

    // Ensure that the organisational unit and user id are valid numbers
    id = GrasshopperUtil.getNumberParam(id);
    userId = GrasshopperUtil.getNumberParam(userId);

    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'Only authenticated users can subscribe to an organisational unit'}).isLoggedInUser(ctx);
    validator.check(id, {'code': 400, 'msg': 'A valid organisational unit id must be provided'}).isInt();
    validator.check(userId, {'code': 400, 'msg': 'A valid user id must be provided'}).isInt();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Ensure the user exists
    UsersDAO.getUser(userId, function(err, user) {
        if (err) {
            return callback(err);
        }

        // Ensure the organisational unit exists
        getOrgUnit(ctx, id, true, function(err, orgUnit) {
            if (err) {
                return callback(err);
            }

            // Ensure the user and organisational unit are on the same application
            if (user.AppId !== orgUnit.AppId) {
                return callback({'code': 401, 'msg': 'The organisational unit must be on the same application as the user'});
            }

            // Ensure the current user can subscribe the specified user to the organisational unit
            OrgUnitAuthz.canSubscribeOrgUnit(ctx, orgUnit, user, function(err, canSubscribe) {
                if (err) {
                    return callback(err);
                } else if (!canSubscribe) {
                    log().warn({'id': orgUnit.id}, 'Unauthorized attempt at subscribing a user to an organisational unit');
                    return callback({'code': 401, 'msg': 'You are not allowed to subscribe a user to an organisational unit'});
                }

                OrgUnitDAO.subscribeOrgUnit(orgUnit, user, callback);
            });
        });
    });
};

/**
 * Unsubscribe a user from the event series and events in an organisational unit
 *
 * @param  {Context}        ctx                 Standard context containing the current user and the current app
 * @param  {Number}         id                  The id of the organisational unit to unsubscribe the user from
 * @param  {Number}         [userId]            The id of the user that should be unsubscribed. Defaults to the current user
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error object, if any
 */
var unsubscribeOrgUnit = module.exports.unsubscribeOrgUnit = function(ctx, id, userId, callback) {
    // Default to the current user if no user was specified
    if (!userId && ctx.user) {
        userId = ctx.user.id;
    }

    // Ensure that the organisational unit and user id are valid numbers
    id = GrasshopperUtil.getNumberParam(id);
    userId = GrasshopperUtil.getNumberParam(userId);

    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'Only authenticated users can unsubscribe from an organisational unit'}).isLoggedInUser(ctx);
    validator.check(id, {'code': 400, 'msg': 'A valid organisational unit id must be provided'}).isInt();
    validator.check(userId, {'code': 400, 'msg': 'A valid user id must be provided'}).isInt();
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Ensure the user exists
    UsersDAO.getUser(userId, function(err, user) {
        if (err) {
            return callback(err);
        }

        // Ensure the organisational unit exists
        getOrgUnit(ctx, id, true, function(err, orgUnit) {
            if (err) {
                return callback(err);
            }

            // Ensure the current user can unsubscribe the specified user from the organisational unit
            OrgUnitAuthz.canUnsubscribeOrgUnit(ctx, orgUnit, user, function(err, canUnsubscribe) {
                if (err) {
                    return callback(err);
                } else if (!canUnsubscribe) {
                    log().warn({'id': orgUnit.id}, 'Unauthorized attempt at unsubscribing a user from an organisational unit');
                    return callback({'code': 401, 'msg': 'You are not allowed to unsubscribe user from an organisational unit'});
                }

                OrgUnitDAO.unsubscribeOrgUnit(orgUnit, user, callback);
            });
        });
    });
};

/* Export */

/**
 * Walk the tree under an organisational unit and export all the data. The export will include
 * the organisational unit and its children. Any event series and/or events associated to the
 * organisational unit or its children will be exported as well
 *
 * @param  {Context}        ctx                 Standard context containing the current user and the current app
 * @param  {Number}         id                  The id of the organisational unit to export
 * @param  {String}         [format]            The format in which the organisational units, event series and events should be exported. Options are `json` or `csv`, defaults to `json`
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error object, if any
 * @param  {Object|String}  callback.data       The exported organisational units, event series and events
 */
var exportOrgUnit = module.exports.exportOrgUnit = function(ctx, id, format, callback) {
    format = format || 'json';

    // Ensure that the organisational unit is a valid number
    id = GrasshopperUtil.getNumberParam(id);

    var validator = new Validator();
    validator.check(null, {'code': 401, 'msg': 'Only authenticated users can export an organisational unit'}).isLoggedInUser(ctx);
    validator.check(id, {'code': 400, 'msg': 'A valid organisational unit id must be provided'}).isInt();
    validator.check(format, {'code': 400, 'msg': 'A valid format must be provided'}).isIn(['csv', 'json']);
    if (validator.hasErrors()) {
        return callback(validator.getFirstError());
    }

    // Ensure the organisational unit exists
    getOrgUnit(ctx, id, false, function(err, orgUnit) {
        if (err) {
            return callback(err);
        }

        // Ensure the current user can export the organisational unit
        OrgUnitAuthz.canExportOrgUnit(ctx, orgUnit, function(err, canExport) {
            if (err) {
                return callback(err);
            } else if (!canExport) {
                log().warn({'id': orgUnit.id}, 'Unauthorized attempt at exporting an organisational unit');
                return callback({'code': 401, 'msg': 'You are not allowed to export the organisational unit'});
            }

            OrgUnitDAO.exportOrgUnit(orgUnit, format, callback);
        });
    });
};
