// gh-orgunit/lib/authz.js
/**
 * Copyright (c) 2015 "Fronteer LTD"
 * Grasshopper Event Engine
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

var GroupsDAO = require('gh-groups/lib/internal/dao');

/**
 * Determine whether or not the user in context can create an organisational unit
 *
 * @param  {Context}    ctx                         Standard context object containing the current user and the current application
 * @param  {Number}     appId                       The id of the application on which to create the organisational unit
 * @param  {OrgUnit}    [parent]                    The organisational unit under which the new organisational unit should be created. If this is left `null`, an authz check for a top-level organisational unit will be created
 * @param  {Function}   callback                    Standard callback function
 * @param  {Object}     callback.err                An error that occurred, if any
 * @param  {Boolean}    callback.canCreate          `true` if the user in context has the appropriate permission. `false` otherwise
 */
var canCreateOrgUnit = module.exports.canCreateOrgUnit = function(ctx, appId, parent, callback) {
    // Anonymous users cannot create organisational units
    if (!ctx.user) {
        return callback(null, false);

    // App administrators (or global admins) can create organisational units
    } else if (ctx.user.canAdmin(appId)) {
        return callback(null, true);

    // At this point, we're dealing with regular users. A regular user can only
    // create an organisational unit under an organisational unit that they can
    // manage.
    } else if (!parent) {
        return callback(null, false);
    }

    // Ensure the user can manage the parent
    return canUpdateOrgUnit(ctx, parent, callback);
};

/**
 * Determine whether or not the user in context can update an organisational unit
 *
 * @param  {Context}    ctx                         Standard context object containing the current user and the current application
 * @param  {OrgUnit}    orgUnit                     The organisational unit to update
 * @param  {Function}   callback                    Standard callback function
 * @param  {Object}     callback.err                An error that occurred, if any
 * @param  {Boolean}    callback.canUpdate          `true` if the user in context has the appropriate permission. `false` otherwise
 */
var canUpdateOrgUnit = module.exports.canUpdateOrgUnit = function(ctx, orgUnit, callback) {
    // Anonymous users cannot update organisational units
    if (!ctx.user) {
        return callback(null, false);

    // App administrators (or global admins) can update organisational units
    } else if (ctx.user.canAdmin(orgUnit.AppId)) {
        return callback(null, true);
    }

    // At this point we need to check if the user is a member of the group
    return GroupsDAO.isMember(orgUnit.Group || orgUnit.GroupId, ctx.user, callback);
};

/**
 * Determine whether or not the user in context can delete an organisational unit
 *
 * @param  {Context}    ctx                         Standard context object containing the current user and the current application
 * @param  {OrgUnit}    orgUnit                     The organisational unit to delete
 * @param  {Function}   callback                    Standard callback function
 * @param  {Object}     callback.err                An error that occurred, if any
 * @param  {Boolean}    callback.canDelete          `true` if the user in context has the appropriate permission. `false` otherwise
 */
var canDeleteOrgUnit = module.exports.canDeleteOrgUnit = function(ctx, orgUnit, callback) {
    return canUpdateOrgUnit(ctx, orgUnit, callback);
};

/**
 * Determine whether or not the user in context can add series to an organisational unit
 *
 * @param  {Context}    ctx                         Standard context object containing the current user and the current application
 * @param  {Number}     orgUnit                     The organisational unit where the user wants to add series to
 * @param  {Function}   callback                    Standard callback function
 * @param  {Object}     callback.err                An error that occurred, if any
 * @param  {Boolean}    callback.canAddSeries       `true` if the user in context has the appropriate permission. `false` otherwise
 */
var canAddOrgUnitSeries = module.exports.canAddOrgUnitSeries = function(ctx, orgUnit, callback) {
    return canUpdateOrgUnit(ctx, orgUnit, callback);
};

/**
 * Determine whether or not the user in context can delete series from an organisational unit
 *
 * @param  {Context}    ctx                         Standard context object containing the current user and the current application
 * @param  {Number}     orgUnit                     The organisational unit where the user wants to delete series from
 * @param  {Function}   callback                    Standard callback function
 * @param  {Object}     callback.err                An error that occurred, if any
 * @param  {Boolean}    callback.canDeleteSeries    `true` if the user in context has the appropriate permission. `false` otherwise
 */
var canDeleteOrgUnitSeries = module.exports.canDeleteOrgUnitSeries = function(ctx, orgUnit, callback) {
    return canUpdateOrgUnit(ctx, orgUnit, callback);
};

/**
 * Determine whether or not the user in context can subscribe a user
 * to the event series and events in an organisational unit
 *
 * @param  {Context}    ctx                         Standard context object containing the current user and the current application
 * @param  {OrgUnit}    orgUnit                     The organisational unit to subscribe a user to
 * @param  {User}       user                        The user to subscribe to the event series and events in an organisational unit
 * @param  {Function}   callback                    Standard callback function
 * @param  {Object}     callback.err                An error that occurred, if any
 * @param  {Boolean}    callback.canSubscribe       `true` if the user in context has the appropriate permission. `false` otherwise
 */
var canSubscribeOrgUnit = module.exports.canSubscribeOrgUnit = function(ctx, orgUnit, user, callback) {
    // Anonymous users cannot subscribe anyone to an organisational unit
    if (!ctx.user) {
        return callback(null, false);

    // Application admins or global admins can subscribe users to organisational units
    } else if (ctx.user.canAdmin(orgUnit.AppId) && ctx.user.canAdmin(user.AppId)) {
        return callback(null, true);

    // Regular users can only subscribe themselves to an organisational unit
    } else if (ctx.user.id === user.id) {
        return callback(null, true);
    }

    // Regular users cannot subscribe anyone else to an organisational unit
    return callback(null, false);
};

/**
 * Determine whether or not the user in context can unsubscribe a user
 * from the event series and events in an organisational unit
 *
 * @param  {Context}    ctx                         Standard context object containing the current user and the current application
 * @param  {OrgUnit}    orgUnit                     The organisational unit to unsubscribe a user from
 * @param  {User}       user                        The user to unsubscribe from the event series and events in an organisational unit
 * @param  {Function}   callback                    Standard callback function
 * @param  {Object}     callback.err                An error that occurred, if any
 * @param  {Boolean}    callback.canUnsubscribe     `true` if the user in context has the appropriate permission. `false` otherwise
 */
var canUnsubscribeOrgUnit = module.exports.canUnsubscribeOrgUnit = function(ctx, orgUnit, user, callback) {
    // Unsubscribing from the event series and events in an organisational unit
    // requires the same permission as subscribing to them
    return canSubscribeOrgUnit(ctx, orgUnit, user, callback);
};

/**
 * Determine whether or not the user in context can export the organisational units, event series and events in an organisational unit
 *
 * @param  {Context}    ctx                         Standard context object containing the current user and the current application
 * @param  {OrgUnit}    orgUnit                     The organisational unit to export
 * @param  {Function}   callback                    Standard callback function
 * @param  {Object}     callback.err                An error that occurred, if any
 * @param  {Boolean}    callback.canExport          `true` if the user in context has the appropriate permission. `false` otherwise
 */
var canExportOrgUnit = module.exports.canExportOrgUnit = function(ctx, orgUnit, callback) {
    // Anonymous users cannot export an organisational unit
    if (!ctx.user) {
        return callback(null, false);
    }

    // Only application admins or global administrators can export an organisational unit
    return callback(null, ctx.user.canAdmin(orgUnit.AppId));
};
