// gh-auth/lib/oauth/dao.js

var _ = require('lodash');

var DB = require('gh-core/lib/db');
var DBUtil = require('gh-core/lib/db/util');
var log = require('gh-core/lib/logger').logger('gh-auth-oauth');

/**
 * Get a page of clients that belong to the provided app
 *
 * @param  {App}            app                         The App whose clients to fetch
 * @param  {Object}         [opts]                      Optional arguments
 * @param  {Number}         [opts.limit]                The maximum number of clients to fetch
 * @param  {Number}         [opts.offset]               The offset index at which to start returning clients
 * @param  {Function}       callback                    Standard callback function
 * @param  {Object}         callback.err                An error that occurred, if any
 * @param  {Object}         callback.response           The query response
 * @param  {Number}         callback.response.count     The total number of clients in the app
 * @param  {OauthClient[]}  callback.response.rows      The clients in this page of results
 */
var getClients = module.exports.getClients = function(app, opts, callback) {
    var options = {
        'where': {'AppId': app.id},
        'offset': opts.offset,
        'limit': opts.limit
    };

    DB.OauthClient.findAndCountAll(options).complete(function(err, clients) {
        if (err) {
            log().error({'err': err, 'appId': app.id}, 'Failed to page OAuth Clients');
            return callback({'code': 500, 'msg': err.message});
        }

        return callback(null, clients);
    });
};

/**
 * Get a client by its id
 *
 * @param  {Number}         id                  The id of the client
 * @param  {Function}       callback            Standard callback function
 * @param  {Object}         callback.err        An error that occurred, if any
 * @param  {OauthClient}    callback.client     The client associated to the given id
 */
var getClient = module.exports.getClient = function(id, callback) {
    DBUtil.findBy('OauthClient', 'id', id, null, callback);
};

var createClient = module.exports.createClient = function(app, displayName, secret, redirectUri, callback) {
    var clientHash = {
        'displayName': displayName,
        'secret': secret,
        'redirectUri': redirectUri,
        'AppId': app.id
    };

    DBUtil.create('OauthClient', clientHash, callback);
};

var updateClient = module.exports.updateClient = function(client, fields, callback) {
    fields = _.pick(fields, 'displayName', 'redirectUri');
    if (_.isEmpty(fields)) {
        return callback({'code': 400, 'msg': 'Attempted to perform a blank update of a client'});
    }

    client.updateAttributes(fields).complete(function(err, client) {
        if (err) {
            log().error({'err': err, 'id': client.id, 'fields': fields}, 'Failed to update a client');
            return callback({'code': 500, 'msg': err.message});
        }

        return callback(null, client);
    });
};

var disableClient = module.exports.disableClient = function(client, callback) {
    if (client.disabled) {
        return callback();
    }

    client.updateAttributes({'disabled': Date.now()}).complete(function(err, client) {
        if (err) {
            log().error({'err': err, 'id': client.id}, 'Failed to disable a client');
            return callback({'code': 500, 'msg': err.message});
        }

        return callback(null, client);
    });
};

var enableClient = module.exports.enableClient = function(client, callback) {
    if (!client.disabled) {
        return callback();
    }

    client.updateAttributes({'disabled': null}).complete(function(err, client) {
        if (err) {
            log().error({'err': err, 'id': client.id}, 'Failed to enable a client');
            return callback({'code': 500, 'msg': err.message});
        }

        return callback(null, client);
    });
};

var setClientSecret = module.exports.setClientSecret = function(client, secret, callback) {
    client.updateAttributes({'secret': secret}).complete(function(err, client) {
        if (err) {
            log().error({'err': err, 'id': client.id}, 'Failed to set a client secret');
            return callback({'code': 500, 'msg': err.message});
        }

        return callback(null, client);
    });
};

var getAuthCode = module.exports.getAuthCode = function(id, callback) {
    DBUtil.findBy('OauthAuthCode', 'id', id, null, callback);
};

var getAuthCodeByCode = module.exports.getAuthCodeByCode = function(code, callback) {
    DBUtil.findBy('OauthAuthCode', 'code', code, null, callback);
};

var createAuthCode = module.exports.createAuthCode = function(client, user, code, expires, callback) {
    var authCodeHash = {
        'code': code,
        'expires': expires,
        'OauthClientId': client.id,
        'UserId': user.id
    };

    DBUtil.create('OauthAuthCode', authCodeHash, callback);
};

/**
 * Get a page of access tokens that belong to the provided user
 *
 * @param  {User}                   user                                    The User whose access tokens to fetch
 * @param  {Object}                 [opts]                                  Optional arguments
 * @param  {Number}                 [opts.limit]                            The maximum number of access tokens to fetch
 * @param  {Number}                 [opts.offset]                           The offset index at which to start returning access tokens
 * @param  {Function}               callback                                Standard callback function
 * @param  {Object}                 callback.err                            An error that occurred, if any
 * @param  {Object}                 callback.response                       The query response
 * @param  {Number}                 callback.response.count                 The total number of access tokens owned by the user
 * @param  {OauthAccessToken[]}     callback.response.rows                  The access tokens in this page of results
 * @param  {OauthClient}            callback.response.rows[i].OauthClient   The client associated to the access token
 */
var getAccessTokensByUser = module.exports.getAccessTokens = function(user, opts, callback) {
    var query = {
        'where': {'UserId': user.id},
        'offset': opts.offset,
        'limit': opts.limit,
        'include': [
            {'model': DB.OauthClient}
        ]
    };

    DB.OauthAccessToken.findAndCountAll(query).complete(function(err, accessTokens) {
        if (err) {
            log().error({'err': err, 'userId': user.id}, 'Failed to page OAuth Access Tokens');
            return callback({'code': 500, 'msg': err.message});
        }

        return callback(null, accessTokens);
    });
};

/**
 * Get an access token by its id
 *
 * @param  {Number}             id                              The id of the access token
 * @param  {Function}           callback                        Standard callback function
 * @param  {Object}             callback.err                    An error that occurred, if any
 * @param  {OauthAccessToken}   callback.accessToken            The access token associated to the given id
 * @param  {OauthClient}        callback.accessToken.Client     The client associated to the access token
 * @param  {User}               callback.accessToken.User       The user who owns the access token
 */
var getAccessToken = module.exports.getAccessToken = function(id, callback) {
    var opts = {
        'include': [
            {'model': DB.OauthClient},
            {'model': DB.User}
        ]
    };

    DBUtil.findBy('OauthAccessToken', 'id', id, opts, callback);
};

/**
 * Get an access token by its unique token string
 *
 * @param  {String}             token                           The unique token string of the access token
 * @param  {Function}           callback                        Standard callback function
 * @param  {Object}             callback.err                    An error that occurred, if any
 * @param  {OauthAccessToken}   callback.accessToken            The access token associated to the given token
 * @param  {OauthClient}        callback.accessToken.Client     The client associated to the access token
 * @param  {User}               callback.accessToken.User       The user who owns the access token
 */
var getAccessTokenByToken = module.exports.getAccessTokenByToken = function(token, callback) {
    var opts = {
        'include': [
            {'model': DB.OauthClient},
            {'model': DB.User}
        ]
    };

    DBUtil.findBy('OauthAccessToken', 'token', token, opts, callback);
};

var getAccessTokensByUser = module.exports.getAccessTokensByUser = function(user, opts, callback) {
    opts = opts || {};
    var query = {
        'where': {
            'UserId': user.id
        },
        'offset': opts.offset,
        'limit': opts.limit,
        'include': [
            {'model': DB.OauthClient}
        ]
    };

    DB.OauthAccessToken.findAndCountAll(query).complete(function(err, response) {
        if (err) {
            log().error({'err': err, 'userId': user.id}, 'Failed to fetch access tokens by user');
            return callback({'code': 500, 'msg': err.message});
        }

        return callback(null, response);
    });
};

var getAccessTokenByClientAndUser = module.exports.getAccessTokenByClientAndUser = function(client, user, callback) {
    var query = {
        'where': {
            'OauthClientId': client.id,
            'UserId': user.id
        }
    };

    DB.OauthAccessToken.find(query).complete(function(err, accessToken) {
        if (err) {
            log().error({
                'err': err,
                'clientId': client.id,
                'userId': user.id
            }, 'Failed to fetch AccessToken by Client and User');
            return callback({'code': 500, 'msg': err.message});
        } else if (!accessToken) {
            log().debug({'clientId': client.id, 'userId': user.id}, 'Attempted to fetch non-existing access token');
            return callback({'code': 404, 'msg': 'Access token not found'});
        }

        return callback(null, accessToken);
    });
};

var createAccessToken = module.exports.createAccessToken = function(client, user, token, callback) {
    var accessTokenHash = {
        'token': token,
        'OauthClientId': client.id,
        'UserId': user.id
    };

    DBUtil.create('OauthAccessToken', accessTokenHash, callback);
};

var deleteAccessToken = module.exports.deleteAccessTokenByClientAndUser = function(accessToken, callback) {
    accessToken.destroy().complete(function(err) {
        if (err) {
            log().error({'err': err}, 'Failed to delete an access token');
            return callback({'code': 500, 'msg': err.message});
        }

        log().debug({'accessToken': accessToken}, 'Deleted an access token');
        return callback();
    });
};
