// gh-admins/lib/api.js
/**
 * Copyright (c) 2014 "Fronteer LTD"
 * Grasshopper Event Engine
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

var AuthUtil = require('gh-auth/lib/util');
var DB = require('gh-core/lib/db');
var GrasshopperUtil = require('gh-core/lib/util');
var log = require('gh-core/lib/logger').logger('gh-users');

var AdminsDAO = require('./internal/dao');

/**
 * Get all global administrators
 *
 * @param  {Number}             [limit]                     The maximum number of results to retrieve. Default: 10
 * @param  {Number}             [offset]                    The paging number of the results to retrieve
 * @param  {Function}           callback                    Standard callback function
 * @param  {Object}             callback.err                An error object, if any
 * @param  {GlobalAdminList}    callback.globalAdmins       All available global administrators
 */
var getGlobalAdmins = module.exports.getGlobalAdmins = function(limit, offset, callback) {
    // TODO: Check that user is a global adminstrator

    // Ensure that the paging values are valid
    limit = GrasshopperUtil.getNumberParam(limit, 10, 1, 25);
    offset = GrasshopperUtil.getNumberParam(offset, 0, 0);

    AdminsDAO.getGlobalAdmins(limit, offset, callback);
};

/**
 * Get a global administrator
 *
 * @param  {Number}             id                          The id of the global administrator to retrieve
 * @param  {Function}           callback                    Standard callback function
 * @param  {Object}             callback.err                An error object, if any
 * @param  {GlobalAdmin}        callback.globalAdmin        The requested global administrator
 */
var getGlobalAdmin = module.exports.getGlobalAdmin = function(id, callback) {
    // TODO: Check that user is a global adminstrator
    // TODO: id validation

    AdminsDAO.getGlobalAdmin(id, callback);
};

/**
 * Get a global administrator by username
 *
 * @param  {String}             username                    The username of the global administrator to retrieve
 * @param  {Function}           callback                    Standard callback function
 * @param  {Object}             callback.err                An error object, if any
 * @param  {GlobalAdmin}        callback.globalAdmin        The requested global administrator
 */
var getGlobalAdminByUsername = module.exports.getGlobalAdminByUsername = function(username, callback) {
    // TODO: Check that user is a global adminstrator
    // TODO: username validation

    AdminsDAO.getGlobalAdminByUsername(username, callback);
};

/**
 * Create a new global administrator
 *
 * @param  {String}             username                    The username for the global administrator
 * @param  {String}             password                    The password for the global administrator
 * @param  {String}             displayName                 The name of the global administrator
 * @param  {Function}           callback                    Standard callback function
 * @param  {Object}             callback.err                An error object, if any
 * @param  {GlobalAdmin}        callback.globalAdmin        The created global administrator
 */
var createGlobalAdmin = module.exports.createGlobalAdmin = function(username, password, displayName, callback) {
    // TODO: Check that user is a global adminstrator
    // TODO: username validation
    // TODO: displayName validation
    // TODO: password validation

    // Ensure that a global administrator with the provided username doesn't alrady exist
    getGlobalAdminByUsername(username, function(err, globalAdmin) {
        if (globalAdmin) {
            log().error({'err': err, 'username': username}, 'Failed to create a new global administrator');
            return callback({'code': 400, 'msg': 'A global administrator with the provided username already exists'});
        } else if (err && err.code !== 404) {
            return callback(err);
        }

        AdminsDAO.createGlobalAdmin(username, password, displayName, callback);
    });
};

/**
 * Update a global administrator
 *
 * @param  {Number}             id                          The id of the global administrator to update
 * @param  {String}             displayName                 Updated global administrator name
 * @param  {Function}           callback                    Standard callback function
 * @param  {Object}             callback.err                An error object, if any
 * @param  {GlobalAdmin}        callback.globalAdmin        The updated global administrator
 */
var updateGlobalAdmin = module.exports.updateGlobalAdmin = function(id, displayName, callback) {
    // TODO: Check that user is a global adminstrator
    // TODO: displayName validation

    // Verify that the provided global administrator exists
    getGlobalAdmin(id, function(err, globalAdmin) {
        if (err) {
            return callback(err);
        }

        AdminsDAO.updateGlobalAdmin(globalAdmin, displayName, callback);
    });
};

/**
 * TODO
 */
var updateGlobalAdminPassword = module.exports.updateGlobalAdminPassword = function(globalAdmin, password, callback) {

};

/**
 * TODO
 */
var deleteGlobalAdmin = module.exports.deleteGlobalAdmin = function(id, callback) {

};

/**
 * Validate the login credentials for a global administrator
 *
 * @param  {String}             username                    The username of the global administrator to attempt login validation for
 * @param  {String}             password                    The password to attempt login validation with
 * @param  {Function}           callback                    Standard callback function
 * @param  {Object}             callback.err                An error object, if any
 * @param  {GlobalAdmin}        callback.globalAdmin        The global administrator for the provided username and password. If the provided password was incorrect, this will be `null`
 */
var validateLoginCredentials = module.exports.validateLoginCredentials = function(username, password, callback) {
    // TODO: username validation
    // TODO: password validation

    // Ensure that a global administrator with the provided username exist
    getGlobalAdminByUsername(username, function(err, globalAdmin) {
        // No global administrator could be found for the provided username
        if (err && err.code === 404) {
            log().info({'err': err, 'username': username}, 'The provided global administrator username is incorrect');
            return callback({'code': 401, 'msg': 'The provided global administratorusername is incorrect'});
        // A different error occured
        } else if (err) {
            log().error({'err': err, 'username': username}, 'Failed to validate the login credentials for a global administrator');
            return callback(err);
        }

        AuthUtil.comparePassword(password, globalAdmin.password, function(err, match) {
            if (err) {
                log().error({'err': err}, 'Failed to validate the password for a global administrator');
                return callback(err);
            } else if (!match) {
                log().info({'err': err, 'username': username}, 'The provided global administrator password is incorrect');
                return callback({'code': 401, 'msg': 'The provided global administrator password is incorrect'});
            }

            return callback(null, globalAdmin);
        });
    });
};
