// gh-admins/lib/api.js
/**
 * Copyright (c) 2014 "Fronteer LTD"
 * Grasshopper Event Engine
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

var Sequelize = require('sequelize');

var DB = require('gh-core/lib/db');
var log = require('gh-core/lib/logger').logger('gh-users');

/**
 * Get all global administrators
 *
 * @param  {Number}             [items]                     The maximum number of results to retrieve. Default: 10
 * @param  {Number}             [page]                      The paging number of the results to retrieve
 * @param  {Function}           callback                    Standard callback function
 * @param  {Object}             callback.err                An error object, if any
 * @param  {GlobalAdminUser[]}  callback.globalAdminUsers   All available global administrators
 */
var getGlobalAdminUsers = module.exports.getGlobalAdminUsers = function(items, page, callback) {

};

/**
 * Get a global administrator by username
 *
 * @param  {String}             username                    The username of the global administrator to retrieve
 * @param  {Function}           callback                    Standard callback function
 * @param  {Object}             callback.err                An error object, if any
 * @param  {GlobalAdminUser}    callback.globalAdminUser    The requested global administrator
 */
var getGlobalAdminUserByUsername = module.exports.getGlobalAdminUserByUsername = function(username, callback) {
    // TODO: Check that user is a global adminstrator
    // TODO: username validation

    AdminsDAO.getGlobalAdminUser(username, callback);
};

/**
 * Create a new global administrator
 *
 * @param  {String}             username                    The username for the global administrator
 * @param  {String}             password                    The password for the global administrator
 * @param  {String}             displayName                 The name of the global administrator
 * @param  {Function}           callback                    Standard callback function
 * @param  {Object}             callback.err                An error object, if any
 * @param  {GlobalAdminUser}    callback.globalAdminUser    The created global administrator
 */
var createGlobalAdminUser = module.exports.createGlobalAdminUser = function(username, password, displayName, callback) {
    // TODO: Check that user is a global adminstrator
    // TODO: username validation
    // TODO: displayName validation
    // TODO: apssword validation

    // Ensure that a global administrator with the provided username doesn't alrady exist
    getGlobalAdminUser(username, function(err, globalAdminUser) {
        if (!(err && err.code === 404)) {
            log().error({'err': err, 'username': username}, 'Failed to create a new global administrator');
            return callback({'code': 500, 'msg': 'A global administrator with the provided username already exists'});
        }

        AdminsDAO.createGlobalAdminUser(username, password, displayName, callback);
    });
};

var updateGlobalAdminUser = module.exports.updateGlobalAdminUser = function(globalAdminUser, displayName, callback) {

};

var updateGlobalAdminUserPassword = module.exports.updateGlobalAdminUserPassword = function(globalAdminUser, password, callback) {

};
